Add Public API spec for Toast Notifications (#1675)
microsoft/WindowsAppSDK: specs/PushNotifications/ToastNotifications-spec.md

should be able to click on a toast popup to launch an app in the correct context. For example, a
news article related toast is expected to launch the News app along with the article in question in
the foreground. Another actionable scenario is for the user to actually interact with the contents
of the toast popup like a UI control button. For example, a messaging app like Teams that displays a
message from another user could have a "Respond To" textbox and button so that the user can directly
respond to the message in the popup without having to launch the application. This scenario triggers
a background process on behalf of the application (with no UI) which inturn processes the "Reply To"
message and forwards the response to the other device.

Toast messages could have a local or cloud based origin. In the case of local toasts, the message
always originates from the app installed on the device. For cloud toasts, the message always
originates from a remote application service that targets the locally installed app in question.

Here is a visual representation of a Simple Toast with no interactive controls:

> ![Toast Simple](https://docs.microsoft.com/en-us/windows/apps/design/shell/tiles-and-notifications/images/send-toast-02.png)

Reviewer: don't use /en-us links - remove that part of the url so that the link opens in the user's language.

(here and below)

Author: Thanks for pointing out. Fixed the links.

The following feature set will not be prioritized for the future include:

-   Object Model to compose Toasts without having to rely on Xml templates.
-   Support for toast collections.

# Examples

## Packaged App Toast Registration

This scenario is specifically geared towards Packaged apps that need to send a toast. An app
Developer Registers their app for Toast Notifications and sets it up for background\foreground
activation. Adding a ToastActivated event handler signals that the app is ready to accept and handle
an action in the foreground where the user clicks on a Toast in the action centre or a UI element in
the toast body (like a Button). The developer can inturn run some business logic in their app in
response to the Toast Action.

```cpp
int main()
{
    auto activationInfo = ToastActivationInfo::CreateFromActivationGuid(winrt::guid(c_comServerId));
    ToastNotificationManager::Default().RegisterActivator(activationInfo);

    auto args = AppInstance::GetCurrent().GetActivatedEventArgs();
    auto kind = args.Kind();

    if (kind == ExtendedActivationKind::Launch)
    {
        // App is launched in Foreground. So intercept toast activators via Foreground event

>         const auto token = ToastNotificationManager::Default().ToastActivated([](const auto&, ToastActivatedEventArgs const& toastActivatedEventArgs)

Reviewer: this line (and others) uses both east and west const - choose one, be constistent [sic] ;)

Author: Fair enough!

        {
                ProcessToastArgs(toastActivatedEventArgs);
        });

        // App does Foreground Stuff Here

        // Cleanup
        ToastNotificationManager::Default().ToastActivated(token);
    }
    else if (kind == ExtendedActivationKind::ToastNotification)
    {
        auto toastActivatedEventArgs = args.Data().as<ToastActivatedEventArgs>();
        ProcessToastArgs(toastActivatedEventArgs);
    }

    return 0;
}
```

## Unpackaged app Toast Registration

Similar to packaged apps, unpackaged apps can also Register themselves to send toasts and act as
targets for toast activation events. There is one basic difference in the Registration for
Unpackaged apps versus their packaged counterparts: they will need to specify their own friendly App
DisplayName and IconUri since they don't have an app manifest file that contains the same.

```cpp
int main()
{

>     Uri absoluteFilePath(L"c:\\test\\icon.png");

Reviewer: Use raw string literals:
```suggestion
    Uri absoluteFilePath(LR"(C:\test\icon.png)");
```

//          hint-inputId="textBox"
//          activationType="background"
//          arguments="action=reply&amp;threadId=92187" />
//  </actions>
//</toast>

const winrt::hstring c_toastLaunchAction = L"action=openThread&amp;threadId=92187";
const winrt::hstring c_toastReplyButtonAction = L"action=reply&amp;threadId=92187";

void ProcessToastArgs(ToastActivatedEventArgs const& toastActivatedEventArgs)
{
    if (toastActivatedEventArgs.ActivationArgs() == c_toastLaunchAction)
    {
        // The user clicks on the toast: So use the launchAction to do stuff
        // Do LaunchAction Stuff
    }
    else if (toastActivatedEventArgs.ActivationArgs() == c_toastReplyButtonAction)
    {
        // The user clicked on the reply button on the toast. So query the input field
        auto input = toastActivatedEventArgs.UserInput();
        auto replyBoxText = input.Lookup(L"ReplyBox");

        // Process the reply text
        ProcessReply(replyBoxText);
    }
}
```

## Processing Toast History

The ToastHistory component is always used to perform Delete and Get operations on behalf of the app.

```cpp
// Get a List of all toasts from the Action Centre

> winrt::Windows::Foundation::IAsyncOperation<Windows::Foundation::Collections::IVector<ToastNotification> > GetToastListASync()

Reviewer: Modern compilers don't require separating the template closing `>` anymore, so you can write:
```suggestion
winrt::Windows::Foundation::IAsyncOperation<Windows::Foundation::Collections::IVector<ToastNotification>> GetToastListASync()
```

{
    auto toastHistory = ToastNotificationManager::Default().History();
    auto toasts = co_await toastHistory.GetAllAsync();
    co_return toasts;
}

// Remove a prior toast using tag and group
winrt::Windows::Foundation::IAsyncAction RemoveToast(const winrt::hstring& tag, const winrt::hstring& group)
{
    auto toastHistory = ToastNotificationManager::Default().History();
    co_await toastHistory.RemoveWithTagGroupAsync(tag, group);
}
```

## Toast Progress Updates

Sometimes a developer would like to show progress bar related updates in a toast:

![Toast Progress](https://docs.microsoft.com/en-us/windows/apps/design/shell/tiles-and-notifications/images/toast-progressbar-annotated.png)

To accomplish that, the developer will need to use the ToastProgressData construct.

```cpp
// Send first Toast Progress Update
void SendUpdatableToastWithProgress()
{
    winrt::hstring tag = L"weekly-playlist";
    winrt::hstring group = L"downloads";

>     winrt::hstring payload = L"<toast launch=\"action = viewDownload &amp; downloadId = 9438108\">"
>                                 L"<visual>"
>                                     L"<binding template = \"ToastGeneric\">"
>                                         L"<text>Downloading this week's new music...</text>"
>                                         L"<progress"
>                                             L"title = \"{progressTitle}\""
>                                             L"value = \"{progressValue}\""
>                                             L"valueStringOverride = \"{progressValueString}\""
>                                             L"status = \"{progressStatus}\" />"
>                                     L"</binding>"
>                                 L"</visual>"
>                             L"</toast>";

Reviewer: ```suggestion
    winrt::hstring payload = LR"(
      <toast launch="action = viewDownload &amp; downloadId = 9438108">
        <visual>
          <binding template = "ToastGeneric">
            <text>Downloading this week's new music...</text>
            <progress
               title="{progressTitle}"
               value="{progressValue}"
               valueStringOverride="{progressValueString}"
               status="{progressStatus}" />
           </binding>
         </visual>
       </toast>";
```

    data.Title(L"Weekly playlist"); // Binds to {progressTitle} in xml payload
    data.Value(0.6); // Binds to {progressValue} in xml payload
    data.ValueStringOverride(L"15/26 songs"); // Binds to {progressValueString} in xml payload
    data.Status(L"Downloading..."); // Binds to {progressStatus} in xml payload
    data.SequenceNumber(1);

    toast.ProgressData(data);
    ToastNotificationManager::Default().ShowToast(toast);
}

// Send subsequent progress updates
winrt::Windows::Foundation::IAsyncAction UpdateProgressAsync()
{
    winrt::hstring tag = L"weekly-playlist";
    winrt::hstring group = L"downloads";

    // Assign new values
    // Note that you only need to assign values that changed. In this example we don't assign progressStatus since we don't need to change it
    ToastProgressData data;
    data.Value(0.7); // Binds to {progressValue} in xml payload
    data.ValueStringOverride(L"18/26 songs"); // Binds to {progressValueString} in xml payload
    data.SequenceNumber(2);

    auto result = co_await ToastNotificationManager::Default().UpdateToastProgressDataAsync(data, tag, group);
    if (result != ToastProgressResult::Succeeded)
    {

>         LOG_HR_MSG(E_UNEXPECTED, "Toast Progress Update Failed!");

Reviewer: it would be good if UpdateToastProgressDataAsync returned something more useful than succeeded / not succeeded. Can it be a struct with an hr? or some way to get back to the error context...

Author: We kind of mapped the behavior in WinAppSDK to mirror the Windows SDK return values for the Progress operation. So we would like to be consistent here since exposing internal HRESULT's may lead to a whole new can of worms.

    // The Result for a Toast Progress related operation
    enum ToastProgressResult
    {
        Succeeded, // The progress operation succeeded
        Failed, // The progress operation failed
        NotificationNotFound, // The progress operation failed to find a toast to process updates
    };

Reviewer 2: DISCUSSION: Succeeded = was updated.
NotificationNotFound = valid call, but notification is gone.
Any other error = exception (app messed up).
We can RoOriginateError with useful diagnostic strings.

## Manifest Registration

For MSIX, the COM activator GUID and the exe need to be registered in the manifest. The launch args
would need to be pre-set to a well-known string that defines Toast Activation Triggers.

```xml
<Extensions>
  <com:Extension Category="windows.comServer">
    <com:ComServer>
      <com:ExeServer Executable="SampleBackgroundApp\SampleBackgroundApp.exe" DisplayName="SampleBackgroundApp" Arguments="----ToastActivationServer">
        <com:Class Id="{GUIDEntryPoint}" />
      </com:ExeServer>
    </com:ComServer>
  </com:Extension>
</Extensions>
```

# API Details

```c#

// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

namespace Microsoft.Windows.ToastNotifications
{
    [experimental]
    // The Shell asset details for Unpackaged App Registrations
    runtimeclass ToastAssets
    {
        // Initialize using Shell assets like DisplayName and iconPath
        ToastAssets(String displayName, Windows.Foundation.Uri iconPath);

        // The App friendly DisplayName for the toast in Action Centre
        String DisplayName { get; };

>         // The full file path for the icon image

Reviewer: can this be an http uri, or just a file / ms-appx kind of path?

Author: It needs to be a file path for unpackaged applications. So something like C:\test.png

Author: ms-appx is not applicable to unpackaged apps unfortunately.

Reviewer: this API supports both unpackaged and packaged apps, though, right? In which case, ms-appx would/should still work?
Using Uri as the type, sort of implies that this is supported, as well as being able to use http/https links

Author: For packaged Apps, we expect the developer to call into: 
static ToastActivationInfo CreateFromActivationGuid(Guid taskClsid);
to retrieve the ToastInfo construct.

For unpackaged, we want the developer to call into:
static ToastActivationInfo CreateFromToastAssets(ToastAssets assets);
to retrieve the ToastInfo construct

The reason for this is that packaged apps already have their assets registered in their app manifest. An unpackaged app like notepad.exe doesn't necessarily have this and has to provide the assets programatically.

Author: That call is followed by RegisterActivator(ToastActivationInfo details); which is true for both Packaged and Unpackaged applications.

>         Windows.Foundation.Uri IconPath { get; };

Reviewer: name seems inconsistent with other places where we have this concept, recommend `IconUri` instead

        ToastWithInlineImage,
    };

    [experimental]
    // The Result for a Toast Progress related operation
    enum ToastProgressResult
    {
        Succeeded, // The progress operation succeeded
        Failed, // The progress operation failed
        NotificationNotFound, // The progress operation failed to find a toast to process updates
    };

    [experimental]
    enum ToastPriority
    {
        Default, // The notification should have default behavior in terms of delivery and display priority during connected standby mode.
        High, // The notification should be treated as high priority. For desktop PCs, this means during connected standby mode the incoming notification can turn on the screen for Surface-like devices if it doesn't have a closed lid detected.
    };

    [experimental]
    // Represent a toast Notification construct
    runtimeclass ToastNotification
    {
        // Initialize a new Toast using an XML Payload.
        ToastNotification(Windows.Data.Xml.Dom.XmlDocument payload);

        // Unique identifier used to replace a notification within a group.
        String Tag;

        // Unique identifier for a toast group in the app

>         String Group;

Reviewer: GroupId to be consistent with ToastId?
or rename ToastId to Id

Author: Renamed to Id

        // A unique identifier for the Toast generated by the platform.
        UInt32 ToastId;

        // The notification Xml Payload
        Windows.Data.Xml.Dom.XmlDocument Payload{ get; };

        // Gets or sets additional information about the toast progress.
        ToastProgressData ProgressData;

        // Gets or sets the time after which a toast notification should not be displayed.

>         Windows.Foundation.DateTime ExpirationTime;

Reviewer: rename to Expiration?

        // Indicates whether the toast will remain in the Action Center after a reboot.
        Boolean ExpiresOnReboot;

        // Gets or sets the priority for a Toast.
        // Hints on how and at what urgency level a notification should be presented to the user (whether to wake up the screen, etc).
        ToastPriority Priority;

        // Gets or sets whether a toast's pop-up UI is displayed on the user's screen.
        Boolean SuppressDisplay;
    };

    [experimental]
    // Supports Toast related operations for all prior displayed Toasts in Action Centre
    runtimeclass ToastNotificationHistory
    {
        // Removes a specific toast with a specific toastIdentifier from Action Centre

>         Windows.Foundation.IAsyncAction RemoveWithIdentiferAsync(UInt32 toastIdentifier);

Reviewer: toastId to be consistent with ToastId above

        // Removes a toast having a specific tag
        Windows.Foundation.IAsyncAction RemoveWithTagAsync(String tag);

        // Removes a toast having a specific tag and group
        Windows.Foundation.IAsyncAction RemoveWithTagGroupAsync(String tag, String group);

        // Remove all toasts for a specific group
        Windows.Foundation.IAsyncAction RemoveGroupAsync(String group);

        // Removes all the toasts for the App from Action Centre
        Windows.Foundation.IAsyncAction RemoveAllAsync();

        // Gets all the toasts for the App from Action Centre

>         Windows.Foundation.IAsyncOperation<Windows.Foundation.Collections.IVector<ToastNotification> > GetAllAsync();

Reviewer: nit, `>>` is ok

Author: Unfortunately MIDL doesnt like it. 
Severity	Code	Description	Project	File	Line	Suppression State
Error	MIDL2025	[msg]syntax error [context]: expecting > or , near ">>"	WindowsAppRuntime_DLL	D:\reunion\ProjectReunion\dev\ToastNotifications\ToastNotifications.idl	174

    };

    [experimental]
    // The manager class which encompasses all Toast API Functionality
    runtimeclass ToastNotificationManager
    {
        // Gets a Default instance of a ToastNotificationManager
        static ToastNotificationManager Default{ get; };

        // Register an activator using an ActivationInfo context and caches the token for unregister
        void RegisterActivator(ToastActivationInfo details);

        // Unregisters the activator and removes the cached Registration token.
        void UnregisterActivator();

        // Event handler for Toast Activations
        event Windows.Foundation.EventHandler<ToastActivatedEventArgs> ToastActivated;

>         // Displays the Toast in Action Centre

Reviewer: nit, Action Center to be consistent with the rest of the docs using en-us spelling

        void ShowToast(ToastNotification toast);

        // Updates the Toast for a Progress related operation using Tag and Group
        Windows.Foundation.IAsyncOperation<ToastProgressResult> UpdateToastProgressDataAsync(ToastProgressData data, String tag, String group);

        // Updates the Toast for a Progress related operation using Tag
        Windows.Foundation.IAsyncOperation<ToastProgressResult> UpdateToastProgressDataAsync(ToastProgressData data, String tag);

        // Get the Toast Setting status for the app
        ToastNotificationSetting Setting{ get; };

        // Gets an instance of ToastHistory
        ToastNotificationHistory History{ get; };

>         // Gets an Xml Payload based ona predefined Toast Template

Reviewer: typo:
```suggestion
        // Gets an Xml Payload based on a predefined Toast Template
```

        Windows.Data.Xml.Dom.XmlDocument GetXmlTemplateContent(ToastTemplateType type);
    };
}
```

# Appendix

-   To support cloud toasts, the developer will need to register the Toast Activator in
    addition to the Push Activator registrations that are already being done. The Windows SDK will

>     inturn figure out the complexity of mapping cloud based toasts to the correct Toast Activator.

Reviewer: typo
```suggestion
    in turn figure out the complexity of mapping cloud-based toasts to the correct Toast Activator.
```

> -   Since building Toast XML payloads during runtime is non-trivial, we encourage developers to
>     utilize the windows developer toolkit, more specifically the ToastContentBuilder APIs to
>     construct the xml payload in the short term. Alternatively they can also use the ToastTemplates that WinAppSDK will provide to build their own toasts.

Reviewer: ```suggestion
-   Since building Toast XML payloads at runtime is non-trivial, we encourage developers to
    utilize the [Windows Community Toolkit](https://docs.microsoft.com/windows/communitytoolkit/), specifically the [ToastContentBuilder](https://docs.microsoft.com/dotnet/api/microsoft.toolkit.uwp.notifications.toastcontentbuilder) APIs to
    construct the XML payload in the short term. Alternatively, they can also use the ToastTemplates that the Windows App SDK will provide to build their own toasts.
```

Desktop applications include Unpackaged apps and Packaged win32 (MSIX Desktop Bridge, Sparse Signed
Packages). The SDK implementation will ensure that all Notification scenarios will adhere to OS
resource management policies like Power Saver, Network Attribution (reporting the amount of data an
app uses), Group Policies, etc. The SDK will also abstract away the complexities of dealing with
Notification delivery and related activations as much as possible freeing the developer to focus on
other app related challenges.

The App Notification API allows Packaged and Unpackaged apps to:

-   Use common APIs to Register and Unregister Notifications.
-   Use common APIs to display Cloud Sourced Notifications.
-   Use common APIs to display Local Notifications.
-   Use common APIs to Update/Delete Notifications.
-   Use AppLifecycle constructs and event handlers to process invoke operations on behalf of the
    user.

# Examples

## Registering and handling a notification

Packaged and Unpackaged Desktop apps will need to register themselves to handle App Notifications.
For Packaged apps, WinAppSDK will use the CLSID defined in the appx manifest to register the current

> process as the COM Server for Notification invokes. The process.exe declared in the manifest will

Reviewer 3: "The process.exe" is strange, how about "The process module (.exe file)"

Author: Renamed.

have to be the same process calling into Register() for the API call to succeed. For Unpackaged

> apps, WinAppSDK will call into Shell libraries to determine appropriate DisplayName and and Icon.

Reviewer 3: "call into shell libraries" is not needed, this is implementation detail.

Author: done

The current process.exe will be the COM Server for Notification invokes.

```cpp

> int main()

Reviewer 3: many calls to `winrt::AppNotificationManager::Default()`. seems better to store the result of that in a temporary to simplify the code and be more efficient.

Author: Done

{

>     // Register an event handler before calling Register()

Reviewer 3: describe this in the positive sense, describing what is desired instead of what is being avoided.
// To ensure all notification handling happens in this process instance register for
// NotificationInvoked before calling Register(). Without this a new process will
// be launched to handle the notification.

    // If this is not done before calling Register(), the OS will bring up a new process everytime for a Notification Invoke.
    const auto token = winrt::AppNotificationManager::Default().NotificationInvoked([](const auto&, const winrt::AppNotificationActivatedEventArgs& notificationActivatedEventArgs)
    {
        ProcessNotificationArgs(notificationActivatedEventArgs);
    });

    winrt::AppNotificationManager::Default().Register();

    auto args = winrt::AppInstance::GetCurrent().GetActivatedEventArgs();
    auto kind = args.Kind();

    if (kind == winrt::ExtendedActivationKind::Launch)
    {
        // App does Launch Stuff Here
    }
    else if (kind == winrt::ExtendedActivationKind::AppNotification)
    {
        auto notificationActivatedEventArgs = args.Data().as<winrt::AppNotificationActivatedEventArgs>();
        ProcessNotificationArgs(notificationActivatedEventArgs);
    }

    // Cleanup event handler
    winrt::AppNotificationManager::Default().NotificationInvoked(token);

>     // Call Unregister() before exiting main so that subsequent invocations will launch a new process

Author: I imagine that during certain scenarios, the developer would want to call this if they prefer a new process launch instead of using the existing process as the COM server. So maybe not really necessary for this scenario where we bail out of main.

>     // Call Unregister() before exiting main so that subsequent invocations will launch a new process

Reviewer 4: If the app crashes, the COM object is destroyed for the current app and any subsequent activations will launch a new instance of the app.

If the app calls Unregister(), any activations will create a new process as the com server.

    winrt::AppNotificationManager::Default().Unregister();
    return 0;
}
```

## Displaying an App Notification

> To display a Notification, an app needs to define a payload using an xml schema. In the example

Reviewer 3: "using an xml schema" -> "in xml". it is assumed that XML conforms to a specific schema.

below, the developer wants to do a bunch of things: Replace a prior chat notification from a friend
for a specific chat group. The developer also wants to make sure that the notification is deleted in
case the user reboots the machine.

```cpp

> void DisplayNotificationForFriend(const winrt::hstring const& payload, const winrt::hstring const& friendId, const winrt::hstring const& chatGroupId)

Reviewer 3: drop the extra `const` in these. C++ WinRT likes east const, so `winrt::hstring const&`

{
    winrt::AppNotification notification(payload);

    // Setting GroupId here ensures that the Notification is only replaced for a specific chat group
    notification.Group(chatGroupId);

    // Setting friend id here ensures that the Notification only replaces the one from this specific friend
    notification.Tag(friendId);

    // Developer sets the Expiry property on reboot to true to delete the notification on reboot
    notification.ExpiresOnReboot(true);
    winrt::AppNotificationManager::Default().Show(notification);
}
```

## Processing a user generated Action on the Notification

The app will need to process the invocation in response to a User interacting with the Notification
in Action Centre. There are 2 common scenarios here after the invoke takes place:

> 1. The developer can choose to have the app launch in in a specific UI context OR

Reviewer 3: "in in" -> "in"

needs to be run without any special UI rendering.

Note: activationType="background" in the Notification payload body is ignored for Desktop apps. It
is upto the developer to process the Activation Arguments and decide whether to display UI or not.

```cpp
//<toast launch="action=openThread&amp;threadId=92187">
//  <visual>
//      <binding template="ToastGeneric">
//          <text hint-maxLines="1">Jill Bender</text>
//          <text>Check out where we camped last weekend! Incredible!!</text>
//          <image placement="hero" src="https://unsplash.it/360/180?image=1043" />
//      </binding>
//  </visual>
//  <actions>
//      <input id="textBox" type="text" placeHolderContent="reply" / >
//      <action
//          content="Send"
//          imageUri="Assets/Icons/send.png"
//          hint-inputId="textBox"
//          activationType="background"
//          arguments="action=reply&amp;threadId=92187" />
//  </actions>
//</toast>

void ProcessNotificationArgs(const winrt::AppNotificationActivatedEventArgs& notificationActivatedEventArgs)
{
    // If the user clicks on a toast, the code will need to launch the chat thread window

>     if (std::wstring(notificationActivatedEventArgs.ActivationArgs().c_str()).find(L"openThread") != std::wstring::npos)

Reviewer 3: store ActivationArgs() in a temporary to simplify the code, use `else if` on line 205

    {
        GenerateChatThreadWindow();
    }
    else // If the user responds to a message by clicking a button in the notification, we will need to reply back to the other user
    if (std::wstring(notificationActivatedEventArgs.ActivationArgs().c_str()).find(L"reply") != std::wstring::npos)
    {
        auto input = notificationActivatedEventArgs.UserInput();
        auto replyBoxText = input.Lookup(L"ReplyBox");

        // Process the reply text
        SendReplyToUser(replyBoxText);
    }
}
```

## Processing Active Notifications

> We define Active Notifications as payloads that have already been displayed by the app and show up

Reviewer 3: "We define..." is unnatural here. and describing a toast as a "payload" is indirect.

> in the Action Centre. In this example, the developer want to delete all Active Notifications from a

Reviewer 3: "developer want" -> "developer wants"

former friend except if the Notification falls under the Archived group.

```cpp

> winrt::Windows::Foundation::IAsyncAction RemoveAllNotificationsFromFormerFriend(std::wstring const& friendId)

Reviewer 3: `FromFormer` is unnecessary

{

>     winrt::AppNotificationManager manager = winrt::AppNotificationManager::Default();

Reviewer 3: `auto manager = winrt::AppNot...` (don't repeat type info)

    auto notifications = co_await manager.GetAllAsync();

>     for (auto notification : notifications)

Reviewer 3: `for (auto& notification : ...)`

    {
        // We remove all notifications with the friend who sent them.The tag is the friendId.
        // We target all possible chat groups where this friend is present except the Archived group
        if (notification.Tag() == friendId && notification.Group() != L"Archived")
        {
            co_await manager.RemoveByTagAndGroupAsync(notification.Tag(), notification.Group());
        }
    }
}
```

## Notification Progress Updates

> Sometimes a developer would like to show progress bar related updates in a Notification:

Reviewer 3: the "a developer" should be removed, be more direct "To show progress ..."

![Notification Progress](https://docs.microsoft.com/windows/apps/design/shell/tiles-and-notifications/images/toast-progressbar-annotated.png)

> To accomplish that, the developer will need to use the AppNotificationProgressData construct to

Reviewer 3: "To accomplish that, use `AppNotificationProgressData` to describe updates to the progress UI presented in the notification"

                <text>Downloading this week's new music...</text>
                <progress
                    title = "{progressTitle}
                    value = "{progressValue}"
                    valueStringOverride = "{progressValueString}"
                    status = "{progressStatus}" />
            </binding>
        </visual>
    </toast>)";

    winrt::AppNotification notification(payload);
    notification.Tag(tag);
    notification.Group(group);

    // Assign initial values for first notification progress UI
    winrt::AppNotificationProgressData data(1);
    data.Title(L"Weekly playlist"); // Binds to {progressTitle} in xml payload
    data.Value(0.6); // Binds to {progressValue} in xml payload
    data.ValueStringOverride(L"15/26 songs"); // Binds to {progressValueString} in xml payload
    data.Status(L"Downloading..."); // Binds to {progressStatus} in xml payload

    notification.Progress(data);
    winrt::AppNotificationManager::Default().Show(notification);
}

// Send subsequent progress updates
winrt::Windows::Foundation::IAsyncAction UpdateProgressAsync()
{
    winrt::hstring tag = L"weekly-playlist";
    winrt::hstring group = L"downloads";

>     // Assign new values

Reviewer 3: remove comment, self-evident

    winrt::AppNotificationProgressData data(2 /* Sequence number */ );
    data.Title(L"Weekly playlist"); // Binds to {progressTitle} in xml payload
    data.Value(0.7); // Binds to {progressValue} in xml payload
    data.ValueStringOverride(L"18/26 songs"); // Binds to {progressValueString} in xml payload
    data.Status(L"Downloading..."); // Binds to {progressStatus} in xml payload

    auto result = co_await winrt::AppNotificationManager::Default().UpdateAsync(data, tag, group);
    if (result == winrt::AppNotificationProgressResult::AppNotificationNotFound)
    {

>         LOG_HR_MSG(E_NOT_SET, "Toast Progress Update Failed since the previous notification update is missing or has been cleared by the user!");

Reviewer 3: logging an error for an expected state is incorrect. consider not using the wil logging or just remove this handling of this case.

    }
}
```

# Remarks

## Registration

> The developer should always call the Notification Registration API first to register the current

Reviewer 3: name the registration method instead of talking about it indirectly. 
the "The developer" is indirect, drop it.

process as the target for Notification invokes. When the app is terminating, the developer should
call Unregister() to free up COM and allow for subsequent invokes to launch a new process.

## Event Handlers

The developer should subscribe to Notification event handlers to handle invocations if the process
happens to be already running. If this isn't done, the OS will launch a new process for every
subsequent invocation.

## Manifest Registration

> For MSIX, the COM activator GUID and the exe need to be specified in the manifest as part of the

Reviewer 3: "MSIX" -> "Packaged applications"

> For MSIX, the COM activator GUID and the exe need to be specified in the manifest as part of the

Reviewer 3: GUID -> CLSID

desktop and com extensions. In the example below, we use an activatorId =
"1920DBA9-0F64-4F0D-8A4B-5E207B812E61". The launch args would need to be pre-set to
"----AppNotificationActivated:" to ensure that WinAppSDK can process the payload as an
AppNotification Kind.

```xml
<Extensions>
    <desktop:Extension Category="windows.toastNotificationActivation">

> 		<desktop:ToastNotificationActivation ToastActivatorCLSID="1920DBA9-0F64-4F0D-8A4B-5E207B812E61" />

Reviewer 3: the indent is wrong here

Author: Ran .prettier and it looks like it resolved this issue.

	</desktop:Extension>
    <com:Extension Category="windows.comServer">
        <com:ComServer>
            <com:ExeServer Executable="SampleBackgroundApp\SampleBackgroundApp.exe" DisplayName="SampleBackgroundApp" Arguments="----AppNotificationActivated:">
                <com:Class Id="1920DBA9-0F64-4F0D-8A4B-5E207B812E61" />
            </com:ExeServer>
        </com:ComServer>
    </com:Extension>
</Extensions>
```

# API Details

```c#
namespace Microsoft.Windows.AppNotifications
{
    // Event args for the Notification Activation
    [experimental]
    runtimeclass AppNotificationActivatedEventArgs
    {
        // Arguments from the invoked button. Empty for Default Activation with no launch args specified in payload.
        String ActivationArgs{ get; };

        // The data from the input elements of a Notification like a TextBox
        Windows.Foundation.Collections.IMap<String, String> UserInput{ get; };
    }

    // Notification Progress Data

>     [experimental]

Reviewer 3: these `[experimental]` attributes need to go.

Author: Removed all experimental attributes. We will do this in our PR as well for Public Preview

> # App Notifications in Windows App SDK

Reviewer 3: in email we talked about the new toast display abilities of Action Center, stacked notifications, etc. those changes have been back ported to Windows 10. some apps will need to detect this. should that be reflected in the API?

Reviewer 3: DISCUSSION: This will be addressed in a future iteration of this design.

# Background

App Notifications are UI popups that that contain rich text, controls and images to
display a message to the user. They are developer driven and originate from a target application

> that is installed by a user on the local device. It is not sufficient for a notification popup on the

Reviewer 5: Below it says they can originate from the network. I presume you mean they "target" or are "associated with" an app on the device? Also adding "local" seems like too much specificity. Suggest maybe "They are developer-driven and are associated with an installed app."

Display notifications and setup their activation handlers: Windows SDK, Win32 Notification Activations via the
Nitro Model and the Windows Toolkit which is a Nuget library. It gets overwhelming for a developer
who does not understand all these technologies to ramp up and get notifications working for their
app.

**Support and Troubleshooting Notifcation Issues**: Because of the API fragmentation mentioned above, it
is also very difficult to diagnose issues when developers hit them and a large portion of developer
cycles are spent diagnosing and troubleshooting problems in various API stacks instead of a single
stack.

**Registration**: The Windows SDK today provides AppId overloads to developers who can specify their
own Identifier for Unpackaged apps. The platform is currently designed to create an explicit
Registration for this overloaded AppId which causes all sorts of problems (string size limits being
hit, Personally Identifiable Information stored as part of AppIds, Registrations tagged with
incorrect App Type etc). The WinApp SDK gets rid of the AppId overloads in the Public APIs.

**Activation**: Setting up activation handlers using Background Triggers is challenging for

> developers because we don't have downlevel support for all OS SKUs. Moreover, the steps for setting

Author: Taken feedback

up activation handlers vary greatly for different app types MSIX Vs Unpackaged Vs UWP. We will need
to abstract away the activation technology and simplify this process for developers for all
supported downlevel OS SKUs.

# Description

This API enables all Win32 applications to display app notifications irrespective
of their app type. This includes unpackaged apps and packaged win32 (MSIX Desktop Bridge, MSIX
Win32App, Sparse Signed Packages). Moreover, all scenarios should adhere to OS resource

> management policies like Power Saver, Network Attribution (amount of data an app uses), enterprise

Author: Correct and we do the work for that in the platform

-   Allow Packaged apps to register for notifications.
-   Allow Unpackaged apps to display cloud sourced notifications.
-   Allow packaged apps to display cloud sourced notifications.
-   Allow packaged apps to display local notifications
-   Allow unpackaged apps to display local notifications
-   Allow packaged apps to Update/Delete their local notifications.
-   Allow unpackaged apps to Update/Delete their local notifications.
-   Allow packaged applications to process notification activators.
-   Allow unpackaged apps to process notification activators.

# Examples

## Packaged App Registering and handling a notification

This scenario is specifically geared towards Packaged apps that need to send an app notification. An app
Developer Registers their app for notifications and sets it up for background\foreground
activation. Adding a AppNotificationActivated event handler signals that the app is ready to accept and handle
an action in the foreground where the user clicks on a Notification in the action centre or a UI element in
the payload body (like a Button). The developer can inturn run some business logic in their app in
response to the UI Action.

```cpp

> CLSID const c_comServerId = { 0x43136EB5,0xD36C,0x11CF,{0xAD,0xBC,0x00,0xAA,0x00,0xA8,0x00,0x33} };

Reviewer 6: @oldnewthing in this case the developer doesn't actually define the COM class themselves (this API does it automagically for them). So they literally do provide a static GUID like provided in the example (which matches the GUID they defined in their app manifest), and the API then creates a COM class that will then call back to the Activated (NotificationInvoked) event handler. Essentially the API abstracts the complexity of COM for the developer, making it be simply a straightforward event handler instead.

> CLSID const c_comServerId = { 0x43136EB5,0xD36C,0x11CF,{0xAD,0xBC,0x00,0xAA,0x00,0xA8,0x00,0x33} };

Reviewer 6: *(And if it's too slow/intensive to parse their manifest, we should consider other ways to validate whether the value is right, Brendan suggested we could look up to see if a COM server is registered with the corresponding CLSID... the important thing to remember is to **provide instantaneous, meaningful feedback to the developer when they make a mistake**... otherwise if they make a typo (which they will), they'll be utterly perplexed why their notifications are silently failing... and developers have wasted hours debugging issues similar to this in our APIs since we don't provide meaningful, immediate errors)*

> CLSID const c_comServerId = { 0x43136EB5,0xD36C,0x11CF,{0xAD,0xBC,0x00,0xAA,0x00,0xA8,0x00,0x33} };

Reviewer 2: DISCUSSION: Register() will be updated so it takes no arguments and it will work for both packaged and unpackaged apps.

For packaged apps, we parse the manifest to get the clsid.

For unpackaged apps, we ask the shell for the app display name and icon.

> CLSID const c_comServerId = { 0x43136EB5,0xD36C,0x11CF,{0xAD,0xBC,0x00,0xAA,0x00,0xA8,0x00,0x33} };

Author: Yes. This was our POR. We will figure out the COM CLSID and validate the launch process as part of the Registration flow.

> CLSID const c_comServerId = { 0x43136EB5,0xD36C,0x11CF,{0xAD,0xBC,0x00,0xAA,0x00,0xA8,0x00,0x33} };

Author: We figured out a way to infer to the Registry and avoid manifest overhead for now.

int main()
{

>     AppNotificationActivationInfo activationInfo{ winrt::guid(c_comServerId) };

Reviewer 5: What is the purpose of the GUID? I would assume the app's PFN would suffice for any kind of identity?

    AppNotificationManager::Default().RegisterActivator(activationInfo);

    auto args = AppInstance::GetCurrent().GetActivatedEventArgs();
    auto kind = args.Kind();

    if (kind == ExtendedActivationKind::Launch)
    {

>         // App is launched in Foreground. So intercept toast activators via Foreground event

Reviewer 7: Nit: maybe rename toast activators to app notification activators?

>         // App is launched in Foreground. So intercept toast activators via Foreground event

Reviewer 2: "Intercept' is the wrong word. We are just handling events.

>         const auto token = AppNotificationManager::Default().AppNotificationActivated([](const auto&, const AppNotificationActivatedEventArgs& notificationActivatedEventArgs)

Reviewer 5: This event is only handled if `args` were `Launch`. But if your launch was for `AppNotification`, the app might keep running and then when the next notification comes in, nothing will happen?

Author: Yes. We will incorporate this feedback as discussed.

>         const auto token = AppNotificationManager::Default().AppNotificationActivated([](const auto&, const AppNotificationActivatedEventArgs& notificationActivatedEventArgs)

Reviewer 6: Quite a verbose API... since the `Activated` event is already off of `AppNotificationManager`, the `AppNotification` portion is a bit redundant, maybe this could simply be...

```cpp
AppNotificationManager::Default().Activated(...)
```

The other APIs don't have this verbosity... showing an AppNotification is just `Show()`, registering the activator is simply `RegisterActivator()`, could we simplify this one too and avoid our APIs running off the horizontal edge of developer's screens? 😊

Author: Incorporated feedback

        {

>             ProcessNotificationArgs(notificationActivatedEventArgs);

Reviewer 2: So we don't care about the Kind that tells us whether we are foreground or background activated?

Author: Incorporated feedback

        });

        // App does Foreground Stuff Here

        // Cleanup
        AppNotificationManager::Default().AppNotificationActivated(token);
    }
    else if (kind == ExtendedActivationKind::AppNotification)
    {
        auto notificationActivatedEventArgs = args.Data().as<AppNotificationActivatedEventArgs>();

>         ProcessNotificationArgs(notificationActivatedEventArgs);

Reviewer 2: Why don't background notifications have to register for the AppNotificationActivated event?

Reviewer 2: DISCUSSION: Raymond didn't undersatnd the API when he wrote this.

    }

    return 0;
}
```

## Unpackaged app Registering and handling a notification

Similar to packaged apps, unpackaged apps can also Register to send app notifications and act as
targets for activation events. There is one basic difference in the Registration for

> Unpackaged apps versus their packaged counterparts: they will need to specify their own friendly App

Reviewer 2: Answer: No. Unpackaged apps always come through launch.

DisplayName and IconUri since they don't have an app manifest file that contains the same.

```cpp
int main()
{
    PCWSTR const c_iconName = L"icon.png";
    std::filesystem::path binaryPath = wil::GetModuleFileNameW<std::wstring>(nullptr);
    std::filesystem::path directoryPath = binaryPath.parent_path();
    std::filesystem::path iconPath = directoryPath / std::filesystem::path(c_iconName);

>     Uri iconUri(iconPath.wstring().c_str());

Reviewer 2: DISCUSSION: Moot. Register() now takes 0 parameters.

>     AppNotificationActivationInfo activationInfo(L"SampleApp", iconUri);

Reviewer 6: Should we consider automatically pulling in the app's display name and icon? When developers write apps for Windows, they don't have to manually specify an icon and display name for Taskbar, that's automatically obtained via their existing assets. Why should we require developers to manually specify display assets for toasts, when we don't require them to manually specify them for taskbar or other UI surfaces?

We could still provide an overload to customize those display assets if desired, but otherwise this feels like it would be creating an opportunity for developers to have inconsistent app branding info since they're now having to manually specify their display assets in another place.

Reviewer 2: DISCUSSION: We're doing it!

Reviewer 8: Will be watching how this is implemented. We currently tap into `IShellItemImageFactory` to show app icons and ... it's not a consistent experience. Hope the override remains in place for folks to specify their own icons and fix their apps when problems arise.

Reviewer 6: @riverar have you used the Community Toolkit Notifications C# library with unpackaged apps today? And have you found problems with it? It automatically obtains display assets, so it'd be good to know if you have problems with that to validate whether your concern is true!

Author: @andrewleader How do we want to address this concern? Should we expose a property off AppNotificationManager for IconUri and DisplayName? We can still keep the single ApNotificationManager::Register() call

>     AppNotificationActivationInfo activationInfo(L"SampleApp", iconUri);

Author: This is not applicable anymore

>     AppNotificationManager::Default().RegisterActivator(activationInfo);

Reviewer 6: I'm not sure the name `RegisterActivator` is the best name for this method, for unpackaged apps, isn't this also registering the app to send notifications, since it's registering the display name and icon URI (two things that have nothing to do with activation and purely have to do with sending/displaying notifications)?

Maybe simply `Register(appInfo)` would work? And then correspondingly, rename `AppNotificationActivationInfo` to `AppNotificationAppInfo` or `AppNotificationRegistrationInfo`, since it's more than just activation info?

Reviewer 9: We will call it Register.

Reviewer 2: DISCUSSION: Single Register() with no parameters.

>     AppNotificationManager::Default().RegisterActivator(activationInfo);

Reviewer 9: Currently the design is the caller of the RegisterActivator is the process that is handling the notifications. Dynamic registering would be something we would consider in the future.

>     AppNotificationManager::Default().RegisterActivator(activationInfo);

Reviewer 2: DISCUSSION: Designed away.

>     AppNotificationManager::Default().RegisterActivator(activationInfo);

Reviewer 2: DISCUSSION: Designed away. Register() takes no parameters any more.

    auto args = AppInstance::GetCurrent().GetActivatedEventArgs();
    auto kind = args.Kind();

    if (kind == ExtendedActivationKind::Launch)
    {
        // App is launched in FG. So intercept toast activators via FG event
        const auto token = AppNotificationManager::Default().AppNotificationActivated([](const auto&, const AppNotificationActivatedEventArgs& notificationActivatedEventArgs)
            {
                ProcessNotificationArgs(notificationActivatedEventArgs);
            });

        // App does Foreground Stuff Here

        // Cleanup
        AppNotificationManager::Default().AppNotificationActivated(token);
    }
    else if (kind == ExtendedActivationKind::AppNotification)
    {

>         auto notificationActivatedEventArgs = args.Data().as<AppNotificationActivatedEventArgs>();

Reviewer 2: >Data [](http://example.com/codeflow?start=51&length=4)

Why does this specific case come in through the args instead of through the AppNotificationActivated event?

Reviewer 2: See discussion earlier.

        ProcessNotificationArgs(notificationActivatedEventArgs);
    }

    return 0;
}
```

## Displaying an App Notification

> To display a notification, an app needs to define a payload using an xml schema. In addition, the developer can set a tag and group property which allows the app to replace a prior notification with the same tag and group with an updated version of the payload.

Reviewer 2: DISCUSSION: Perfect match of both is required.

> To display a notification, an app needs to define a payload using an xml schema. In addition, the developer can set a tag and group property which allows the app to replace a prior notification with the same tag and group with an updated version of the payload.

Reviewer 2: DISCUSSION: Same as UWP. Check that our docs explain what the difference is (if any).

Finally, the ExpiresOnReboot property can also be set to ensure that the notification is automatically deleted from
the Action Centre after device reboots.

```cpp
// Display a notification in action centre using a tag and group

> void DisplayNotification(const winrt::Windows::Data::Xml::Dom::XmlDocument& doc)

Reviewer 5: Does the API rely on `XmlDocument? Even if it's only sample code, we should avoid this type since most developers will be using a different XML parser (or just strings) this is a problem. Moving data between different parsers is inefficient.

Reviewer 9: This is a short term problem we have that we are planning to rectify with a "toast builder" API in the future.

Reviewer 5: Tying this API to the WinRT XML DOM is a big mistake. AFAICT the implementation in Windows just serializes the DOM and then uses XmlLite to parse it -- meaning we could have the WinAppSDK deal in Strings only. Down-level it turns into a DOM to pass to the OS (which then turns it back into a String) but up-level we can update the OS to rely on Strings and save everyone effort.

Reviewer 2: RECOMMEND: Just take a string.

/* static AppNotification.CreateFromXml(string xml); */

AppNotification(String contentXml);

> void DisplayNotification(const winrt::Windows::Data::Xml::Dom::XmlDocument& doc)

Author: I made this scenario specific now!

{
    AppNotification notification(doc);
    // Setting Tag, Group and ExpiresOnReboot properties are optional
    notification.Tag(L"Tag");
    notification.Group(L"Group");
    notification.ExpiresOnReboot(true);
    AppNotificationManager::Default().Show(notification);
}
```
## Processing an Activation

The app will need to process the activator in response to a User interacting with the notification in Action Centre. 2 common
scenarios here are

1. Have the app launch in the foreground in a specific UI context OR
2. Have the app process an action specific behavior (like a button press in the payload body) in the background.

```cpp
//<toast launch="action=openThread&amp;threadId=92187">
//  <visual>
//      <binding template="ToastGeneric">
//          <text hint-maxLines="1">Jill Bender</text>
//          <text>Check out where we camped last weekend! Incredible!!</text>

> //          <image placement="hero" src="https://unsplash.it/360/180?image=1043" />

Author: done!

//      </binding>
//  </visual>
//  <actions>
//      <input id="textBox" type="text" placeHolderContent="reply" / >
//      <action
//          content="Send"
//          imageUri="Assets/Icons/send.png"
//          hint-inputId="textBox"

> //          activationType="background"

Reviewer 9: It's not tied to the activation kind. The arguments property are tied to the ExtendedActivationKind. We're going to follow up with the shell but there is a possibility that "activationType" is used by them for animations and other purposes.

> //          activationType="background"

Reviewer 2: >   activationType="background" [](http://example.com/codeflow?start=9&length=30)

Important thing to call out: This attribute is ignored for Win32 apps!

Author: Added in the description right above the code.

> //          activationType="background"

Reviewer 2: DISCUSSION: This attribute is basically useless in Win32.

> //          activationType="background"

Author: Yes. Put in a note to account to make this clearer.

const winrt::hstring c_notificationReplyButtonAction = L"action=reply&amp;threadId=92187";

void ProcessNotificationArgs(const AppNotificationActivatedEventArgs& notificationActivatedEventArgs)
{
    if (notificationActivatedEventArgs.ActivationArgs() == c_notificationLaunchAction)
    {
        // The user clicks on the notification: So use the launchAction to do stuff
        // Do LaunchAction Stuff
    }
    else if (notificationActivatedEventArgs.ActivationArgs() == c_notificationReplyButtonAction)
    {
        // The user clicked on the reply button on the notification. So query the input field
        auto input = notificationActivatedEventArgs.UserInput();
        auto replyBoxText = input.Lookup(L"ReplyBox");

        // Process the reply text
        ProcessReply(replyBoxText);
    }
}
```

## Processing Active Notifications

We define active notifications as payloads that have already been displayed by the app and show up in the
Action Centre. We can perform a number of operations on these notifications such as retrieval and
deletion.

```cpp
// Get a List of all active notifications displayed in the Action Centre

> winrt::Windows::Foundation::IAsyncOperation<Windows::Foundation::Collections::IVector<AppNotification>> GetNotificationListASync()

Reviewer 3: `Async` as the suffix

Reviewer 2: ASync -> Async

{

>     auto notifications = co_await AppNotificationManager::Default().GetAllAsync();

Reviewer 5: Is there a non-`Async` version?

Reviewer 2: DISCUSSION: No.

Current guidance suggest providing sync options, because people don't like async.

winrt::Windows::Foundation::IAsyncAction RemoveNotification(const winrt::hstring tag, const winrt::hstring group)
{
    co_await AppNotificationManager::Default().RemoveWithTagGroupAsync(tag, group);
}
```

## Notification Progress Updates

Sometimes a developer would like to show progress bar related updates in a notification:

![Notification Progress](https://docs.microsoft.com/windows/apps/design/shell/tiles-and-notifications/images/toast-progressbar-annotated.png)

To accomplish that, the developer will need to use the AppNotificationProgressData construct to update the progress bar on the screen.

```cpp
// Send first Notification Progress Update
void SendUpdatableNotificationWithProgress()
{
    winrt::hstring tag = L"weekly-playlist";
    winrt::hstring group = L"downloads";

    winrt::hstring payload =
    LR"(<toast launch="action = viewDownload &amp; downloadId = 9438108">
        <visual>
            <binding template = "ToastGeneric">
                <text>Downloading this week's new music...</text>
                <progress
                    title = "{progressTitle}
                    value = "{progressValue}"

>                     valueStringOverride = "{progressValueString}"

Reviewer 2: DISCUSSION: Yup this works. 

But you can't say

<text>{progressTitle}</text>

The only place {progressTitle} is allowed is in the value of the attribute "title".

title="....{progressTitle}..."

Also

title="....{progressTitle}...{progressTitle}..."
title="{progressTitle} ({progressValue}% complete)"

are not allowed.

                    status = "{progressStatus}" />
            </binding>
        </visual>
    </toast>)";

    XmlDocument doc;
    doc.LoadXml(payload);
    AppNotification notification(doc);
    notification.Tag(tag);
    notification.Group(group);

    // Assign initial values for first notification progress UI
    AppNotificationProgressData data;
    data.Title(L"Weekly playlist"); // Binds to {progressTitle} in xml payload
    data.Value(0.6); // Binds to {progressValue} in xml payload
    data.ValueStringOverride(L"15/26 songs"); // Binds to {progressValueString} in xml payload

>     data.Status(L"Downloading..."); // Binds to {progressStatus} in xml payload

Reviewer 2: DISCUSSION: Start small.

    data.SequenceNumber(1);

    notification.Progress(data);
    AppNotificationManager::Default().Show(notification);
}

// Send subsequent progress updates
winrt::Windows::Foundation::IAsyncAction UpdateProgressAsync()
{
    winrt::hstring tag = L"weekly-playlist";

>     winrt::hstring group = L"downloads";

Author: Factored them out of the Functions

    // Assign new values

>     // Note that you only need to assign values that changed. In this example we don't assign progressStatus since we don't need to change it

Reviewer 2: DISCUSSION: The comment about "You only need to assign values that changed" may be incorrect. Please investigate and fix.

if it's correct, then we have open questions about how to detect which values have changed.

>     // Note that you only need to assign values that changed. In this example we don't assign progressStatus since we don't need to change it

Author: The developer is expected to refresh each and every property in ProgressData. Otherwise, we assume that the property default (empty string in this case) takes over.

    AppNotificationProgressData data;
    data.Value(0.7); // Binds to {progressValue} in xml payload

>     data.ValueStringOverride(L"18/26 songs"); // Binds to {progressValueString} in xml payload

Reviewer 2: DISCUSSION: Comment may have described feature incorrectly. Please investigate.

>     data.ValueStringOverride(L"18/26 songs"); // Binds to {progressValueString} in xml payload

Author: It needs to be set or the API will end up substituing an empty string if unset by default. Made that clear in the new sample update.

    data.SequenceNumber(2);

    auto result = co_await AppNotificationManager::Default().UpdateProgressDataAsync(data, tag, group);
    if (result == AppNotificationProgressResult::AppNotificationNotFound)
    {

>         LOG_HR_MSG(E_NOT_SET, "Toast Progress Update Failed since the previous notification update is missing or has been cleared by the user!");

Reviewer 5: Can you tell the difference? Being cleared by the user is an expected state. Having the notification not be there is likely an app bug?

Reviewer 5: So is it realistic to log this? People will copy-paste sample code. I think a comment is better than a log here.

Author: Ok just added a comment for now

    }
}
```
# Remarks

## Registration

The developer should always call the Notification Registration API first to register the current process as

> the Activator target.

Author: REmoved activation target in favor of "target for Notification invokes"

## Foreground API calls

The developer should always subscribe to AppNotification activation foreground events to intercept notification related actions if

> the app happens to be running in the foreground.

Author: Reworded to:
The developer should subscribe to Notification event handlers to handle invocations if
the process happens to be already running. If this isn't done, the OS will launch a new process for every subsequent invocation.

## Manifest Registration

For MSIX, the COM activator GUID and the exe need to be registered in the manifest. The launch args

> would need to be pre-set to a well-known string that defines Activation Triggers.

Author: It is a static string that we will document publicly.

Reworded to:
For MSIX, the COM activator GUID and the exe need to be specified in the manifest as part of the desktop and com extensions. The launch args
would need to be pre-set to "----AppNotificationActivated:" to ensure that WinAppSDK can process the payload as an AppNotificationKind.

```xml
<Extensions>

>   <com:Extension Category="windows.comServer">

Author: Yes. Thankyou for catching this. Will update the doc.

    <com:ComServer>

>       <com:ExeServer Executable="SampleBackgroundApp\SampleBackgroundApp.exe" DisplayName="SampleBackgroundApp" Arguments="----AppNotificationActivationServer">

Reviewer 6: We should consider having the notification library parse and validate their Executable path and ensure it matches to the current executable path... Developers could easily mistype their executable, and then be left clueless why activation isn't working if we don't validate things like that.

Reviewer 5: Is it against the rules to specify a different exe? Like maybe you have a light-weight, headless process for handling in-line replies.

Reviewer 6: If the feature team wants to support different EXEs, a parameter in the Register method could be added to specify the different EXE for the 1% of apps that want that additional complexity (and then still ensuring the majority of apps have quality, informative errors... I have prior experience from developers, getting your EXE path incorrect was one of the biggest stumbling blocks developers had when trying to use this before I started automatically validating that in the Toolkit library).

Reviewer 8: I think you're asking for two things:

1. Validation of the path. Seems reasonable.
2. Enforcement of "path must match current running image (or whatever)". Seems unnecessarily restrictive.

Author: For now, we are going to enforce that the COM server path in the manifest matches the current process. I do understand that we want to give the developer a way to override this in the future by perhaps including a Process.exe overload. For example,  Register(String processNameExe). We will evaluate as we keep getting feedback.

Author: Your feedback has been incorporated via a Simple AppNotificationManager::Register() implementation :)

>         <com:Class Id="{GUIDEntryPoint}" />

Author: No it would be an actual GUID. REplacing this with an actual GUID.

      </com:ExeServer>
    </com:ComServer>
  </com:Extension>
</Extensions>
```

# API Details

```c#
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

import "..\AppLifecycle\AppLifecycle.idl";

> namespace Microsoft.Windows.AppNotifications

Reviewer 2: Push notificatoins were M.W.PushNotifications
So this is a natural sibling.

{
    // The Registration Info for Packaged and Unpackaged Notification Activations
    runtimeclass AppNotificationActivationInfo
    {

>         // Initialize using a manifest defined COM Activator Id. Only applicable to Packaged Win32 applications

Author: Designed this away with a simple Register() call.

        AppNotificationActivationInfo(Guid taskClsid);

>         // Initialize using Notification Assets. Only applicable to Unpackaged Win32 applications which need to specify their own assets like DisplayName and Icon.

Reviewer 5: Some of these comments are too long to see in Github (and there's no scroll bar... ?!?). Please wrap :)

>         AppNotificationActivationInfo(String displayName, Windows.Foundation.Uri iconUri);

Reviewer 5: URI implies it could be `ms-appx` or `ms-data` in addition to web etc. If it's just a file path, probably make this a string?

        // The CLSID associated with the Client COM server that Windows App SDK will activate

>         Guid TaskClsid{ get; };

Reviewer 5: How is this populated if the developer uses the second ctor (with name and icon)?

Reviewer 6: Maybe we don't need to create a standalone class for ActivationInfo, and just have the `Register` method take the constructor values directly?

* For packaged apps: `Register(clsid)`
* For unpackaged apps: `Register(displayName, icon)`

Author: This is good feedback. I will go with this for now.

Author: New feedback is to simply call it Register() for both Packaged and Unpackaged apps

        // The App friendly DisplayName for the Notification in Action Centre 

>         String DisplayName{ get; };

Reviewer 2: >DisplayName [](http://example.com/codeflow?start=15&length=11)

How is the unpakaged app identified when it launches a second time, so we can tell it about the notifications from previous instances that are still visible in Action Center? Do the DisplayName+IconUri both have to match? Only the DisplayName? Only the IconUri? Executable path? Do they never match and we assume every instance is a new app unrelated to any previous instance?

Author: We use the Process's full file path. By design: We will lose the notifications if the app decides to change the path in any way.

        // The full file path for the icon image
        Windows.Foundation.Uri IconUri{ get; };
    }

    // Event args for the Notification Activation

>     runtimeclass AppNotificationActivatedEventArgs

Reviewer 10: The same interface (Windows.ApplicationModel.Activation.IActivatedEventArgs) is used from the platform.

>     runtimeclass AppNotificationActivatedEventArgs

Reviewer 2: DISCUSSION: Not needed because this is not the actual activation args. It's the args.Data, which is IInspectable.

    {
        // Arguments from the invoked button. Empty for Default Activation with no launch args specified in payload.

>         String ActivationArgs{ get; };

Author: I'm good calling this just Arguments

>         String ActivationArgs{ get; };

Reviewer 2: DISCUSSION: UWP calls it "Argument" (singular)
RECOMMEND: Argument

>         String ActivationArgs{ get; };

Author: updated

        // The data from the input elements of a Notification like a TextBox
        Windows.Foundation.Collections.IMapView<String, String> UserInput{ get; };
    }

    // Notification Progress Data
    runtimeclass AppNotificationProgressData
    {
        // Initializes a new Instance of NotificationProgressData
        AppNotificationProgressData();

        // Gets or sets the sequence number of this notification data.

>         // When multiple NotificationProgressData objects are received, the system displays the data with the greatest non-zero number. 

Reviewer 5: Nit: no need to say "non-zero."

Reviewer 5: Good point... which makes this property REQUIRED if the default Uint32 of 0 is used. So unless the ctor of this class sets it to 1 (my preference), it will be a bug farm for developers "why didn't my update show?" "Oh I forgot to put '1' in the stupid sequence field."

Reviewer 2: DISCUSSION: Zero sequence number is special and means "ignore".

Should we make it constructor parameter so you can't forget to set it?

RECOMMEND:
AppNotificationProgressData(UInt32 sequenceNumber);

throw if they pass 0.

Author: Done!

        DisabledForUser, // Notification is blocked by a user defined Global Setting
        DisabledByGroupPolicy, // Notification is blocked by Group Policy
        DisabledByManifest, // Notification is blocked by a setting in the manifest. Only for packaged applications.
    };

    // The Result for a Notification Progress related operation
    enum AppNotificationProgressResult
    {
        Succeeded, // The progress operation succeeded
        AppNotificationNotFound, // The progress operation failed to find a Notification to process updates
    };

    // The Priority of the Notification UI associated with it's popup in the Action Centre
    enum AppNotificationPriority
    {
        Default, // The notification should have default behavior in terms of delivery and display priority during connected standby mode.
        High, // The notification should be treated as high priority. For desktop PCs, this means during connected standby mode the incoming notification can turn on the screen for Surface-like devices if it doesn't have a closed lid detected.
    };

    runtimeclass AppNotification
    {

>         AppNotification(Windows.Data.Xml.Dom.XmlDocument payload);

Reviewer 5: Please no :). We should default to a `String` constructor. Have a static `CreateFromXml` if we need it. (Relying on the WinRT DOM was one an unfortunate mistake in the original WinRT notifications API).

>         // Unique identifier used to replace a notification within a group.

Author: The existing tag group hierarchy that developers are already comfortable with in the Windows SDK is being applied here as well. There is no change in the API hierarchy.

        String Tag;

        // Unique identifier for a Notification group in the app
        String Group;

        // A unique identifier for the Notification generated by the platform.

>         UInt32 Id { get; };

Author: unique per user session. Its simply an increment of the last notificationId starting from 0.

        // The notification Xml Payload

>         Windows.Data.Xml.Dom.XmlDocument Payload{ get; };

Reviewer 5: Again, have a string and add (eg) a `PayloadAsXml` if needed.

Author: Feedback incorporated.

>         Windows.Data.Xml.Dom.XmlDocument Payload{ get; };

Reviewer 5: At what point does the system read the DOM? For example, if I use this object to start a notification and then immediately mutate the DOM, what happens? Has the API taken a deep clone before it returns?

Author: Our API converts the DOM back to String. I think we still need to do a basic validation of the xml schema when we create a new Notification type but use string everywhere else internally.

Author: Incorporated Feedback

        // Gets or sets additional information about the Notification progress.

>         AppNotificationProgressData Progress;

Reviewer 11: If I write to this, when is that write reflected? Properties are supposed to be "very cheap" to read/write. Is there an "Update" method on this or on the Manager that takes a modified AppNotification and updates the live copy?

        // Gets or sets the time after which a Notification should not be displayed.
        Windows.Foundation.DateTime Expiration;

        // Indicates whether the Notification will remain in the Action Center after a reboot.
        Boolean ExpiresOnReboot;

        // Gets or sets the priority for a Notification.

>         // Hints on how and at what urgency level a notification should be presented to the user (whether to wake up the screen, etc).

Reviewer 5: Priority or urgency? They are different concepts; we should use the appropriate terms (in API and in docs).

Author: Priority is consistent with Windows SDK and is the consistent naming convention.

        AppNotificationPriority Priority;

        // Gets or sets whether a Notification's pop-up UI is displayed on the user's screen.

>         Boolean SuppressDisplay;

Reviewer 5: Sounds like it's suppressing the physical display (esp. since the `Priority` talks about turning on the display). I assume it's a "negative" property name so that the default (false) means "show the thing" (double-negative) but you can have a positive name and set the "default" value to true in the constructor. For example, `ShowNotification` or `DisplayNotification` etc. 

More broadly, I'm not sure why you'd send a toast and not show it. Is this so it shows up in Action Center but doesn't annoy me first with the popup? If so, then maybe something like `AddDirectlyToNotificationCenter` or something (not the best name...)

Reviewer 6: The old platform name for this was `SurpressPopup`. Given we seem to talk about these as "popups", was there a reason `SuppressDisplay` was chosen instead? Or `SuppressBanner` since the Settings app calls these "notification banners"? I agree that `Display` is an odd choice.

Author: @jonwis Any thoughts on this point? This feedback was from your prior recommendation. I'm totally ok with rolling back :)

Author: No comments on this. So keeping as is.

    }

    // The manager class which encompasses all App Notification API Functionality
    [experimental]
    runtimeclass AppNotificationManager
    {
        // Gets a Default instance of a AppNotificationManager

>         static AppNotificationManager Default{ get; };

Reviewer 2: >Default [](http://example.com/codeflow?start=38&length=7)

Could this ever return null?

Reviewer 2: DISCUSSION: No.

>         // Register an activator using an ActivationInfo context and caches the token for unregister

Reviewer 5: Caches what token? There's no return value.

Author: Got rid of inside baseball in comments per feedback

>         void RegisterActivator(AppNotificationActivationInfo details);

Reviewer 5: "Activator" sounds like such Microsoft-speak. Maybe just "Handler"?

Author: Incorporated this feedback

>         void RegisterActivator(AppNotificationActivationInfo details);

Reviewer 5: What happens if I call this twice?

Author: It will throw. Only one call to Register is allowed. Developer can still call Unregister and Register again which is supported as a pattern. eg. Games may want to disable notifications during gameplay and can use this pattern to suppress toasts at their end during active gaming sessions.

        // Unregisters the activator and removes the cached Registration token.

>         void UnregisterActivator();

Reviewer 5: Is there a use-case for unregistering? Wouldn't you just unhook your event handlers, and not send any more toasts?

Author: Correct. Unregister tells COM that we are done with this process. Henceforth, a susbsequent activation will launch a new process.

>         void UnregisterActivator();

Reviewer 2: >UnregisterActivator [](http://example.com/codeflow?start=13&length=19)

If I unregister while an AppNotificationAction event is active, does it block until the event handler returns? Or does it unregister immediately and let the event handler continue asynchronously?

        // Event handler for Notification Activations

>         event Windows.Foundation.TypedEventHandler<Microsoft.Windows.AppLifecycle.ExtendedActivationKind, AppNotificationActivatedEventArgs> AppNotificationActivated;

Reviewer 11: First parameter to the event is the _sender_ of the event, so it should probably be `AppNotificationManager`

Author: Feedback incorporated

>         event Windows.Foundation.TypedEventHandler<Microsoft.Windows.AppLifecycle.ExtendedActivationKind, AppNotificationActivatedEventArgs> AppNotificationActivated;

Reviewer 2: >AppNotificationActivated [](http://example.com/codeflow?start=141&length=24)

What thread is this event raised on? Is it raised on the thread that called RegisterActivator? Is it it raised on the apartment that called RegisterActivator? Is it raised on a background thread? Is it raised on an arbitrary thread?

Can two of these events be outstanding simultaneously? Or do we serialize them?

Can this event be raised at the same time the app is dealing with an AppNotification extended activation?

Reviewer 2: DISCUSSION: Incoming RPC thread = arbitrary background thread.

Two events can be active simultaneously.

        // Displays the Notification in Action Center

>         void Show(AppNotification notification);

Reviewer 2: >Show [](http://example.com/codeflow?start=13&length=4)

What happens if I call Show() on a notification that was returned by GetAllAsync()?

Reviewer 2: DISCUSSION: Throws.

        // Updates the Notification for a Progress related operation using Tag and Group

>         Windows.Foundation.IAsyncOperation<AppNotificationProgressResult> UpdateProgressDataAsync(AppNotificationProgressData data, String tag, String group);

Reviewer 2: >UpdateProgressDataAsync [](http://example.com/codeflow?start=74&length=23)

Confusing name, since it doesn't update the ProgressData object!

Author: Feedback incorporated

        // Updates the Notification for a Progress related operation using Tag

>         Windows.Foundation.IAsyncOperation<AppNotificationProgressResult> UpdateProgressDataAsync(AppNotificationProgressData data, String tag);

Reviewer 11: Should this be `AppNotificationProgressUpdateResult` ?

Reviewer 6: Additionally, I'm not sure this should have the name `Progress` in it... while the typical use case for providing updateable data in your notification is for showing a progress bar, that's not the ONLY use case... devs can use this API to silently update text in their notification (could be a countdown rather than a progress indicator, for example). We could just call these `UpdateData` and `AppNotificationData` / etc, simply remove the `Progress` term from it. Or call it `UpdateBindableDataAsync` if we want it to be more clear.... because basically these values can be bound to the UI of the notification, similar to XAML data binding.

>         Windows.Foundation.IAsyncOperation<AppNotificationProgressResult> UpdateProgressDataAsync(AppNotificationProgressData data, String tag);

Reviewer 2: We can remove notifications by Tag, Group, or Tag+Group.

We can update a notification by Tag, Tag+Group, but not Group alone.

Why the mismatch?

Author: Because you can have multiple notifications in the Action Centre today with the same Group. A toast that attempts to replace every item in that group is an unsupported scenario. So the developer is expected to provide: tag, or tag + group for operations like replace or update to work.

        // Get the Notification Setting status for the app

>         AppNotificationSetting Enablement { get; };

Reviewer 5: "Enablement" is a strange word. "Status" might be better, but status of what? `EnabledState` might work, since it's not a bool.

Reviewer 3: Existing Toast API uses `Setting` for the property name. There are a few uses of `Enablement`, perhaps `EnablementPolicy`. `Availability` is another choice, many uses of that in the existing APIs. I recommend biasing towards existing API.

>         AppNotificationSetting Enablement { get; };

Reviewer 2: >Enablement [](http://example.com/codeflow?start=31&length=10)

AccessStatus is what we traditionally call this.

Reviewer 2: DISCUSSION: ToastNotifier calls it Setting.

CONSIDER: Setting

Author: Incorporated feedback

        // Removes a specific Notification with a specific NotificationIdentifier from Action Centre
        Windows.Foundation.IAsyncAction RemoveByIdAsync(UInt32 notificationId);

        // Removes a Notification having a specific tag
        Windows.Foundation.IAsyncAction RemoveWithTagAsync(String tag);

        // Removes a Notification having a specific tag and group
        Windows.Foundation.IAsyncAction RemoveWithTagGroupAsync(String tag, String group);

        // Remove all Notifications for a specific group

>         Windows.Foundation.IAsyncAction RemoveGroupAsync(String group);

Reviewer 2: >RemoveGroupAsync [](http://example.com/codeflow?start=40&length=16)

Why is it Remove*With*TagAsync, but RemoveGroupAsync (no "with")?

Reviewer 2: DISCUSSION:

RemoveById
RemoveByTag
RemoveByGroup
RemoveByTagAndGroup

Author: Feedback incorporated

Here is a visual representation of a notification with a Message "Reply To" option:

![Toast Interactive 2](https://docs.microsoft.com/windows/apps/design/shell/tiles-and-notifications/images/toast-notification.png)

For more details see:

-   [Toast Notification WinRT APIs](https://docs.microsoft.com//uwp/api/Windows.UI.Notifications.ToastNotification?view=winrt-20348)
    Defines all the API constructs that we have for notifications in WinRT today using the
    Windows SDK.
-   [Toast Notification UX Guidance using Toolkit](https://docs.microsoft.com/windows/apps/design/shell/tiles-and-notifications/toast-ux-guidance)
    Defines the UX Guidance for developers who want to display notifications using the toolkit libraries.
-   [Sending Local Toasts using C# using Toolkit](https://docs.microsoft.com/windows/apps/design/shell/tiles-and-notifications/send-local-toast?tabs=uwp)
    Defines how an app developer can send interactive notifications to the user and setup activation
    handlers for these notifications using the Toolkit libraries.

> ## The problems today

Author: That's covered in the Description section below.

**Cloud Notifications integration with Windows App SDK**: While we support Push scenarios for Raw notification
on behalf of unpackaged applications in the Windows App SDK 1.0, we do not support Cloud sourced App Notifications. We need

> to fill this gap for 1.1 and beyond.

Author: This needs to be significantly reworded. Push is actually shipped for 1.1 as Experimental right @vaheeshta ? In light of that, we want to ensure that these APIs fill in the Cloud integration gap. I will reword this statement to this if it is acceptable:

While we support Push scenarios for Raw notification
on behalf of unpackaged applications in the Windows App SDK 1.1 as an experimental feature, we do not support Cloud sourced App Notifications. We need
to fill this gap for future versions.

> to fill this gap for 1.1 and beyond.

Author: The above statement avoids associating this new feature set with a WinAppSDK ship version.

> to fill this gap for 1.1 and beyond.

Reviewer 12: Paw raw notifications for **packaged** apps are available in WinAppSDK 1.0-exp1. I would change to: While we support Push scenarios for Raw notification on behalf of packaged applications in the Windows App SDK, we do not support Cloud sourced App Notifications. We need to fill this gap.

> to fill this gap for 1.1 and beyond.

Author: Yeah. That sounds like a good middle ground.

> **API fragmentation**: There are too many API technologies that are published today to simply

Author: The goal here is that this covers all the cases for publicly distributed apps since this doc is geared towards the public WinAppSDK developer. 

Inbox components are a seperate discussion and can still continue using their choice of technology. Nothing stopping them from using reunion either if they need to.

> **API fragmentation**: There are too many API technologies that are published today to simply

Author: I had some time to process this question since I likely misunderstood it. Yes, the goal here is that the API super sets everything we have. From a notification perspective, it should cover every scenario that we already cover today + some more (as needed). As far as deprecation for inbox APIs are concerned, that would be a north star.

> **API fragmentation**: There are too many API technologies that are published today to simply

Reviewer 3: But I understand WinAppSDK does not cover UWP's. Given that it is not a superset.

> **API fragmentation**: There are too many API technologies that are published today to simply

Author: Yes

is also very difficult to diagnose issues when developers hit them and a large portion of developer
cycles are spent diagnosing and troubleshooting problems in various API stacks instead of a single
stack.

**Registration**: The Windows SDK today provides AppId overloads to developers who can specify their
own Identifier for Unpackaged apps. The platform is currently designed to create an explicit
Registration for this overloaded AppId which causes all sorts of problems (string size limits being
hit, Personally Identifiable Information stored as part of AppIds, Registrations tagged with
incorrect App Type etc). The WinApp SDK gets rid of the AppId overloads in the Public APIs.

**Activation**: Setting up activation handlers using Background Triggers is challenging for
developers because we don't have downlevel support for all OS SKUs. Moreover, the steps for setting
up activation handlers vary greatly for different app types MSIX Vs Unpackaged Vs UWP. We will need
to abstract away the activation technology and simplify this process for developers for all
supported downlevel OS SKUs.

> **Notification Composition**: There are well known gaps in the Windows SDK today in the area of composing

Author: Fair enough. This is in our near roadmap though.

the Notification content via Xml schema. The Windows Toolkit gets around this limitation by providing a well-defined
object model to compose payloads but all developers don't use the Toolkit. We have a plan to build a Windows Toolkit like
NotificationBuilder model in the near future for WinAppSdk.

# Description

> At a high level, we need to provide a way for all Win32 applications to display app notifications irrespective

Author: Updated.

of their app type. This includes unpackaged apps and packaged win32 (MSIX Desktop Bridge, MSIX
Win32App, Sparse Signed Packages). Moreover, all scenarios should adhere to OS resource
management policies like Power Saver, Network Attribution (amount of data an app uses), enterprise
group policies, etc. The WinAppSDK will abstract away the complexities of
dealing with Notification delivery and related activations as much as possible freeing the developer
to focus on other app related challenges.

> We will prioritize the following feature set for Notifications in Windows App SDK:

Author: resolved by changing the wording

-   App Notification Registrations for Unpackaged apps.
-   App Notification Registrations for Packaged apps
-   Cloud sourced App Notifications for Unpackaged Win32 apps.
-   Cloud sourced App Notifications for Packaged Win32 apps.
-   Local App Notification delivery and Create/Update/Delete support for Packaged Win32 apps.
-   Local App Notification delivery and Create/Update/Delete support for unpackaged Win32 apps.

> The following feature set will not be prioritized for the future include:

Author: done

-   Object Model to compose payloads without having to rely on Xml templates.
-   Support for categorizing App Notifications by Collection.

# Examples

> ## Packaged App Notification Registration

Reviewer 2: CONSIDER: Clarify this title to describe what it is doing.

This scenario is specifically geared towards Packaged apps that need to send an app notification. An app
Developer Registers their app for notifications and sets it up for background\foreground
activation. Adding a AppNotificationActivated event handler signals that the app is ready to accept and handle
an action in the foreground where the user clicks on a Notification in the action centre or a UI element in
the payload body (like a Button). The developer can inturn run some business logic in their app in
response to the UI Action.

```cpp
int main()
{

>   auto activationInfo = AppNotificationActivationInfo::CreateFromActivationGuid(winrt::guid(c_comServerId));

Author: Resolved

>   auto activationInfo = AppNotificationActivationInfo::CreateFromActivationGuid(winrt::guid(c_comServerId));

Author: Done

>   auto activationInfo = AppNotificationActivationInfo::CreateFromActivationGuid(winrt::guid(c_comServerId));

Author: Done

    AppNotificationManager::Default().RegisterActivator(activationInfo);

    auto args = AppInstance::GetCurrent().GetActivatedEventArgs();
    auto kind = args.Kind();

    if (kind == ExtendedActivationKind::Launch)
    {
        // App is launched in Foreground. So intercept Notification activators via Foreground event
        const auto token = AppNotificationManager::Default().AppNotificationActivated([](const auto&, const AppNotificationActivatedEventArgs& notificationActivatedEventArgs)
        {
                ProcessNotificationArgs(notificationActivatedEventArgs);
        });

        // App does Foreground Stuff Here

        // Cleanup
        AppNotificationManager::Default().AppNotificationActivated(token);
    }

>     else if (kind == ExtendedActivationKind::AppNotification)

Reviewer 3: is this the background case? only executing if the app is not already running? does this use the -Embedding command line to detect the COM activation?

Author: Correct. This is specifically for the background case. WinAppSDK will parse the command line for -Embedding to detect the activator args.

Reviewer 2: 1. The app will hook up the event handle before registering.

The notification subsystem does not guarantee activation ordering, so there is no "new" race between the event and the activation args.

Why not just use the event for all cases?

DISCUSSION: If the developer never calls Register, then all activations start a new process and go through the "case AppNotification" path.

When the app calls Register, the AppNotificationManager checks if there is any handler registered for AppNotificationActivated. if not, then it assumes the app wants a new activation for every notification.

RECOMMEND: Take offline to reconcile against the principle that registering an event handler does not alter behavior.

>     else if (kind == ExtendedActivationKind::ToastNotification)

Reviewer 7: Do we still keep the type name to be ToastNotification?

Author: Thanks for catching this. Yes. We need to introduce a new kind in AppLifeCycle called AppNotification. Updated.

    {
        auto notificationActivatedEventArgs = args.Data().as<AppNotificationActivatedEventArgs>();
        ProcessNotificationArgs(notificationActivatedEventArgs);
    }

    return 0;
}
```

## Unpackaged app Notification Registration

Similar to packaged apps, unpackaged apps can also Register to send app notifications and act as
targets for activation events. There is one basic difference in the Registration for
Unpackaged apps versus their packaged counterparts: they will need to specify their own friendly App
DisplayName and IconUri since they don't have an app manifest file that contains the same.

```cpp
int main()
{

>     Uri absoluteFilePath(LR"(C:\test\icon.png)");

Author: Done.

>     AppNotificationAssets assets(L"SampleApp", absoluteFilePath);

Author: Ok removed the types in favor of parameters to a constructor

    notification.Group(L"Group");
    notification.ExpiresOnReboot(true);
    AppNotificationManager::Default().Show(notification);
}
```
## Processing an Activation

The app will need to process the activator in response to a User interacting with the notification in Action Centre. 2 common
scenarios here are

1. Have the app launch in the foreground in a specific UI context OR
2. Have the app process an action specific behavior (like a button press in the payload body) in the background.

```cpp
//<toast launch="action=openThread&amp;threadId=92187">
//  <visual>
//      <binding template="ToastGeneric">
//          <text hint-maxLines="1">Jill Bender</text>
//          <text>Check out where we camped last weekend! Incredible!!</text>
//          <image placement="hero" src="https://unsplash.it/360/180?image=1043" />
//      </binding>
//  </visual>
//  <actions>
//      <input id="textBox" type="text" placeHolderContent="reply" / >
//      <action
//          content="Send"
//          imageUri="Assets/Icons/send.png"
//          hint-inputId="textBox"
//          activationType="background"
//          arguments="action=reply&amp;threadId=92187" />
//  </actions>
//</toast>

> const winrt::hstring c_notificationLaunchAction = L"action=openThread&amp;threadId=92187";

Reviewer 3: not very realistic using a constant for `threadId`

Author: I agree. This was meant to simply demonstrate usage. The Launch Args can be any predefined string "foo" that the developer is aware about.

Reviewer 5: Instead you could just check whether it contains `"openThread"` or `"reply"`.

Reviewer 2: SAMPLE

    {
        // The user clicks on the notification: So use the launchAction to do stuff
        // Do LaunchAction Stuff
    }
    else if (notificationActivatedEventArgs.ActivationArgs() == c_notificationReplyButtonAction)
    {
        // The user clicked on the reply button on the notification. So query the input field
        auto input = notificationActivatedEventArgs.UserInput();
        auto replyBoxText = input.Lookup(L"ReplyBox");

        // Process the reply text
        ProcessReply(replyBoxText);
    }
}
```

## Processing Active Notifications

We define active notifications as payloads that have already been displayed by the app and show up in the
Action Centre. We can perform a number of operations on these notifications such as retrieval and
deletion.

```cpp
// Get a List of all active notifications displayed in the Action Centre
winrt::Windows::Foundation::IAsyncOperation<Windows::Foundation::Collections::IVector<AppNotification>> GetNotificationListASync()
{
    auto notifications = co_await AppNotificationManager::Default().GetAllAsync();
    co_return notifications;
}

// Remove a notification from action centre using tag and group

> winrt::Windows::Foundation::IAsyncAction RemoveNotification(const winrt::hstring& tag, const winrt::hstring& group)

Reviewer 3: dangerous use of & param in async method

Author: Removed &

{
    co_await AppNotificationManager::Default().RemoveWithTagGroupAsync(tag, group);
}
```

## Notification Progress Updates

Sometimes a developer would like to show progress bar related updates in a notification:

![Notification Progress](https://docs.microsoft.com/windows/apps/design/shell/tiles-and-notifications/images/toast-progressbar-annotated.png)

To accomplish that, the developer will need to use the AppNotificationProgressData construct to update the progress bar on the screen.

```cpp
// Send first Notification Progress Update
void SendUpdatableNotificationWithProgress()
{
    winrt::hstring tag = L"weekly-playlist";
    winrt::hstring group = L"downloads";

>     winrt::hstring payload = LR"(<toast launch="action = viewDownload &amp; downloadId = 9438108">

Author: Addressed

    AppNotificationProgressData data;
    data.Value(0.7); // Binds to {progressValue} in xml payload
    data.ValueStringOverride(L"18/26 songs"); // Binds to {progressValueString} in xml payload
    data.SequenceNumber(2);

    auto result = co_await AppNotificationManager::Default().UpdateProgressDataAsync(data, tag, group);
    if (result != AppNotificationProgressResult::Succeeded)
    {
        LOG_HR_MSG(E_UNEXPECTED, "Notification Progress Update Failed!");
    }
}
```
# Remarks

## Registration

The developer should always call the Notification Registration API first to register the current process as
the Activator target.

## Foreground API calls

The developer should always subscribe to AppNotification activation foreground events to intercept notification related actions if
the app happens to be running in the foreground.

## Manifest Registration

For MSIX, the COM activator GUID and the exe need to be registered in the manifest. The launch args
would need to be pre-set to a well-known string that defines Activation Triggers.

```xml
<Extensions>
  <com:Extension Category="windows.comServer">
    <com:ComServer>

>       <com:ExeServer Executable="SampleBackgroundApp\SampleBackgroundApp.exe" DisplayName="SampleBackgroundApp" Arguments="----AppNotificationActivationServer">

Reviewer 3: when COM activates a LocalServer it passes [-Embedding command line](https://docs.microsoft.com/en-us/windows/win32/com/localserver32#remarks). is this argument used as well?

the sample does not show using either. is that missing?

Author: The reunion implementation will look for the keyword and parse the arguments accordingly

        <com:Class Id="{GUIDEntryPoint}" />
      </com:ExeServer>
    </com:ComServer>
  </com:Extension>
</Extensions>
```

# API Details

```c#
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

namespace Microsoft.Windows.AppNotifications
{
    [experimental]
    // The Shell asset details for Unpackaged App Registrations 

>     runtimeclass AppNotificationAssets

Reviewer 3: having a class to express 2 params in one factory function unneeded. you can always add more overloads if new params are needed.

Author: Sure. I removed NotificationAssets and included a parameterized ctor for ActivationInfo instead.

    {
        // Initialize using Shell assets like DisplayName and iconPath
        AppNotificationAssets(String displayName, Windows.Foundation.Uri iconUri);

        // The App friendly DisplayName for the Notification in Action Centre 
        String DisplayName { get; };

        // The full file path for the icon image
        Windows.Foundation.Uri IconUri { get; };
    }

    [experimental]
    // The Registration Info for Packaged and Unpackaged Notification Activations
    runtimeclass AppNotificationActivationInfo
    {
        // Initialize using a manifest defined COM Activator Id. Only applicable to Packaged Win32 applications

>         static AppNotificationActivationInfo CreateFromActivationGuid(Guid taskClsid);

Reviewer 3: the clsid is a COM local server CLSID. seems those terms should be used for the factory function, or just use a constructor to avoid having to give this a name.

Author: The issue here is that we would need 2 constructors: One that takes in a CLSID and another that takes in AppNotificationAssets. Overload constructors are disallowed in midl3 the last I checked.

Author: Added 2 ctors. This happens to work now in the idl because the 2 ctors have different number of arguments.

        // Initialize using Notification Assets. Only applicable to Unpackaged Win32 applications which need to specify their own assets like DisplayName and Icon.
        static AppNotificationActivationInfo CreateFromAppNotificationAssets(AppNotificationAssets assets);

        // The CLSID associated with the Client COM server that Windows App SDK will activate
        Guid TaskClsid{ get; };

        // The Shell assets associated with the Unpackaged app
        AppNotificationAssets Assets{ get; };

>     };

Reviewer 3: nit, `;` not needed here

Author: removed everywhere

        // Gets/Sets the Value for the numerical Progress percentile: a number between 0 and 1. Binds to {progressValue} in progress xml tag.
        Double Value;

        // Gets/Sets the Value for the Progress String. Binds to {progressValueString} in progress xml tag
        String ValueStringOverride;

        // Gets/Sets the Value for the Status. Binds to {progressStatus} in progress xml tag
        String Status;
    };

    [experimental]
    // The Notification User Setting or Notification Group Policy Setting
    enum AppNotificationSetting
    {
        Enabled, // Notification is not blocked by settings or group policy
        DisabledForApplication, // Notification is blocked by a user defined App Setting
        DisabledForUser, // Notification is blocked by a user defined Global Setting
        DisabledByGroupPolicy, // Notification is blocked by Group Policy
        DisabledByManifest, // Notification is blocked by a setting in the manifest. Only for packaged applications.
    };

    [experimental]
    // The Result for a Notification Progress related operation
    enum AppNotificationProgressResult
    {
        Succeeded, // The progress operation succeeded

>         Failed, // The progress operation failed

Reviewer 3: what kinds of failures? catastrophic failures should be reported via exceptions.

Author: Thought about this a bit more. 
I believe we just need Succeeded and AppNotificationNotFound. Succeeded tells us that the ProgressUpdate call was successful. NotFound tells the developer that they didn't send the initial progress notification. So the subsequent progressupdate failed.

Everything else is catastrophic and can be exposed via exceptions.

        AppNotificationNotFound, // The progress operation failed to find a Notification to process updates
    };

    [experimental]
    enum AppNotificationPriority
    {
        Default, // The notification should have default behavior in terms of delivery and display priority during connected standby mode.
        High, // The notification should be treated as high priority. For desktop PCs, this means during connected standby mode the incoming notification can turn on the screen for Surface-like devices if it doesn't have a closed lid detected.
    };

    [experimental]

>     // Represent a Notification Notification construct

Reviewer 3: Notification repeated. best to just delete this comment as it add no new information.

    runtimeclass AppNotification
    {

>         // Initialize a new Notification using an XML Payload.

Reviewer 3: self-evident comment, delete

        AppNotification(Windows.Data.Xml.Dom.XmlDocument payload);

        // Unique identifier used to replace a notification within a group.
        String Tag;

        // Unique identifier for a Notification group in the app
        String Group;

        // A unique identifier for the Notification generated by the platform.

>         UInt32 Id;

Reviewer 3: its mutable, seems should be `{get;};` if the platform is the source.

Author: Correct platform is the source. Will make the change.

        Windows.Foundation.DateTime Expiration;

        // Indicates whether the Notification will remain in the Action Center after a reboot.
        Boolean ExpiresOnReboot;

        // Gets or sets the priority for a Notification.
        // Hints on how and at what urgency level a notification should be presented to the user (whether to wake up the screen, etc).
        AppNotificationPriority Priority;

        // Gets or sets whether a Notification's pop-up UI is displayed on the user's screen.
        Boolean SuppressDisplay;
    };

    [experimental]
    // The manager class which encompasses all App Notification API Functionality
    runtimeclass AppNotificationManager
    {
        // Gets a Default instance of a AppNotificationManager
        static AppNotificationManager Default{ get; };

        // Register an activator using an ActivationInfo context and caches the token for unregister
        void RegisterActivator(AppNotificationActivationInfo details);

        // Unregisters the activator and removes the cached Registration token.
        void UnregisterActivator();

        // Event handler for Notification Activations

>         event Windows.Foundation.EventHandler<AppNotificationActivatedEventArgs> AppNotificationActivated;

Reviewer 3: use `TypedEventHandler<ExtendedActivationKind, AppNotificationActivatedEventArgs>` here, EventHandler is reserved for static events. see API guidelines for details

        // Displays the Notification in Action Center
        void Show(AppNotification notification);

        // Updates the Notification for a Progress related operation using Tag and Group
        Windows.Foundation.IAsyncOperation<AppNotificationProgressResult> UpdateProgressDataAsync(AppNotificationProgressData data, String tag, String group);

        // Updates the Notification for a Progress related operation using Tag
        Windows.Foundation.IAsyncOperation<AppNotificationProgressResult> UpdateProgressDataAsync(AppNotificationProgressData data, String tag);

        // Get the Notification Setting status for the app

>         AppNotificationSetting Setting{ get; };

Reviewer 3: Consider Enablement instead of Setting. or split the value into `Boolean Enabled { get; };` and `DisabledReason { NotDisabled, GroupPolicy, etc.};`

Author: Fixed to Enablement

        // Removes a specific Notification with a specific NotificationIdentifier from Action Centre

>         Windows.Foundation.IAsyncAction RemoveWithIdentiferAsync(UInt32 notificationIdentifier);

Reviewer 3: seems `By` is better than `With` here. `Id` instead of `Identifier` to match the field. So `RemoveByIdAsync()`

Author: fixed

# Toast Notifications in Windows App SDK

# Background

> Toast Notifications are UI popups that are added that contain rich text, controls and images to display a message to the user. They are developer driven and originate from a target application that is installed by a user on the local device. It is not sufficient for a toast popup on the screen to simply be displayed to the user, it also needs to be actionable. For example, the user should be able to click on a toast popup to launch an app in the correct context. For example, a news article related toast is expected to launch the News app along  with the article in question in the foreground. Another actionable scenario is for the user to actually interact with the contents of the toast popup like a UI control button. For example, a messaging app like Teams that displays a message from another user could have a "Respond To" textbox and button so that the user can directly respond to the message in the popup without having to launch the application. This scenario triggers a background process on behalf of the application (with no UI) which inturn processes the "Reply To" message and forwards the response to the other device.

Author: I believe the section below: Problems today should cover that.

> Toast Notifications are UI popups that are added that contain rich text, controls and images to display a message to the user. They are developer driven and originate from a target application that is installed by a user on the local device. It is not sufficient for a toast popup on the screen to simply be displayed to the user, it also needs to be actionable. For example, the user should be able to click on a toast popup to launch an app in the correct context. For example, a news article related toast is expected to launch the News app along  with the article in question in the foreground. Another actionable scenario is for the user to actually interact with the contents of the toast popup like a UI control button. For example, a messaging app like Teams that displays a message from another user could have a "Respond To" textbox and button so that the user can directly respond to the message in the popup without having to launch the application. This scenario triggers a background process on behalf of the application (with no UI) which inturn processes the "Reply To" message and forwards the response to the other device.

Author: I went ahead and formatted the doc using prettierrc

Here is a visual representation of a Toast with a Message "Reply To" option:

![Toast Interactive 2](https://docs.microsoft.com/en-us/windows/apps/design/shell/tiles-and-notifications/images/toast-notification.png)

For more details see:

-   [Toast Notification WinRT APIs](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Notifications.ToastNotification?view=winrt-20348)
    Defines all the API constructs that we have for Toast Notifications in WinRT today using the Legacy Windows SDK.
-   [Toast Notification UX Guidance using Toolkit](https://docs.microsoft.com/en-us/windows/apps/design/shell/tiles-and-notifications/toast-ux-guidance)
    Defines the UX Guidance for developers who want to display toasts using the toolkit libraries.
-   [Sending Local Toasts using C# using Toolkit](https://docs.microsoft.com/en-us/windows/apps/design/shell/tiles-and-notifications/send-local-toast?tabs=uwp)
    Defines how an app developer can send interactive toasts to the user and setup activation handlers for these toasts using the Toolkit libraries.

## The problems today

> **Cloud Toast integration with Windows SDK**: While we support Push scenarios for Raw notification on behalf of unpackaged applications in the Windows SDK 1.0, we do not support Cloud Toasts. We need to fill this gap for 1.1 and beyond.

Author: Yes for packaged apps, if we push a toast down, shell knows how to resolve the identity of the toast when it ultimately receives it. For Unpackaged apps, there is no toast registration story. So shell does not know how to resolve the toast and what identity to use.

**API fragmentation**: There are too many API technologies that are published today to simply Display toasts and setup their activation handlers: Windows Legacy SDK, Win32 Toast Activations via the Nitro Model and the Windows Toolkit which is a Nuget library. It gets overwhelming for a developer who does not understand all these technologies to ramp up and get toast support working for their app. 

**Support and Troubleshooting Toast Issues**: Because of the API fragmentation mentioned above, it is also very difficult to diagnose issues when developers hit them and a large portion of developer cycles are spent dignosing and troubleshooting problems in various API stacks instead of a single stack.

**Activation**: Setting up activation handlers using Background Toast Triggers is challenging for developers because we don't have downlevel support for all OS SKUs. Moreover, the steps for setting up activation handlers vary greatly for different app types MSIX Vs Unpackaged Vs UWP. We will need to abstract away the activation technology and simplify this process for developers for all supported downlevel OS SKUs. 

> **Toast Composition**: There are well known gaps in the Legacy Windows SDK today in the area of composing the Toast content. The Windows Toolkit gets around this limitation by providing a well-defined object model to compose toasts but all developers don't use the Toolkit.

Author: Sounds good. I'll just call it Windows SDK

# Description

At a high level, we need to provide a way for all Win32 applications to display toasts irrespective of their app type. This includes unpackaged apps and packaged win32
(MSIX Desktop Bridge, MSIX Win32App, Sparse Signed Packages). Moreover, all Toast scenarios should
adhere to OS resource management policies like Power Saver, Network Attribution (amount of data an
app uses), enterprise group policies, etc. The Windows App SDK Push component will abstract away the
complexities of dealing with Toast delivery and toast related activations as much as
possible freeing the developer to focus on other app related challenges.

We will prioritize the following feature set for Toast Notifications in Windows App SDK:

-   Toast Registrations for Unpackaged apps.
-   Toast Registrations for Packaged apps
-   Cloud sourced Toasts for Unpackaged Win32 apps.
-   Cloud sourced Toasts for Packaged Win32 apps.

> -   Local Toast delivery and CRUD operations for Packaged Win32 apps.

Author: Create/Update/Delete :) But yes, I will be specific and won't use acronyms

-   Local Toasts delivery and CRUD operations for unpackaged Win32 apps.

The following feature set will not be prioritized for the next release but will be prioritized for the second half of 2022:

-   Object Model to compose Toasts without having to rely on Xml templates.
-   Support for Granular toasts and toast collections.

# Examples

## Packaged App Toast Registration

> This scenario is specifically geared towards Packaged apps that need to send a toast. An app Developer Registers their app for Toast Notifications and sets it up for background\foreground activation. 

Author: Hi Jon. Just a small correction here. The ToastActivated event is not for the "Toast being displayed from shell scenario". It is the scenario where the user interacts with a toast in the UI after it is displayed and the app does something in response to that action.

> This scenario is specifically geared towards Packaged apps that need to send a toast. An app Developer Registers their app for Toast Notifications and sets it up for background\foreground activation. 

Author: But i will elaborate more since this needs to be part of the details.

```cpp
// Display a toast with tag and group
void DisplayToast(const winrt::Windows::Data::Xml::Dom::XmlDocument& doc)
{
    ToastNotification toast = winrt::make<ToastNotification>(doc);
    // Setting Tag, Group and ExpiresOnReboot properties are optional
    toast.Tag(L"Tag");
    toast.Group(L"Group");
    toast.ExpiresOnReboot(true);
    ToastNotificationManager::ShowToast(toast);
}
```

## Create a Toast Payload using Preexisting Toast Xml templates

We will provide a small set of templates that developers can use to compose payloads. The developer can then use the Document Object Model (DOM) to customize the content in the payload. Below is an example of such customization using a simple payload with Image and Text. Here the developer adds an attribute to the image element and sets the text fields in the payload.

```cpp
// Uses an API template to Compose and Display a toast
void CreateToastPayload()
{

>     auto doc = ToastNotificationManager::GetXmlTemplateContent(ToastTemplateType::ToastImageAndText01);

Author: @vaheeshta  just updated the templates with more descriptive titles. Hopefully the template names now can get the point across. There is also a nice fancy table above showing how the notification would look with each template. Many thanks to her for defining this!

>     auto images = doc.GetElementsByTagName(L"image");

Author: For the next release, we want to replace templates with ToastBuilder construct. That way, no more xml for the developer to worry about. So something like: ToastBuilder.AddTitle("Foo").AddText("Bar").AddImage("c:\foo.png").AddButton(button);

>     auto images = doc.GetElementsByTagName(L"image");

Author: @nicuparente for future plans. We really want to get to this soon though : )

    auto srcAttribute = doc.CreateAttribute(L"src");
    srcAttribute.Value(L"images/toastImageAndText.png");

>     images.GetAt(0).Attributes().SetNamedItem(srcAttribute);

Author: auto images = doc.GetElementsByTagName(L"image");
3 lines above. But I can clump the two lines together if that makes this easier to read.

    auto textNodes = doc.GetElementsByTagName(L"text");

>     for (auto textNode : textNodes)

Author: Thanks. Will fix.

## Manifest Registration

For MSIX, the COM activator GUID and the exe need to be registered in the manifest. The launch args
would need to be pre-set to a well-known string that defines Toast Activation Triggers.

```xml
<Extensions>
  <com:Extension Category="windows.comServer">
    <com:ComServer>
      <com:ExeServer Executable="SampleBackgroundApp\SampleBackgroundApp.exe" DisplayName="SampleBackgroundApp" Arguments="----ToastActivationServer">
        <com:Class Id="{GUIDEntryPoint}" />
      </com:ExeServer>
    </com:ComServer>
  </com:Extension>
</Extensions>
```

# API Details

```c#

// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

namespace Microsoft.Windows.ToastNotifications
{
    [experimental]
    // The Shell asset details for Unpackaged App Registrations 
    runtimeclass ToastAssets
    {
        // Initialize using Shell assets like DisplayName, iconPath and icon Background color
        ToastAssets(String displayName, String iconPath, Windows.UI.Colors iconBackgroundColor);

>         // Initialize using Shell assets like DisplayName and iconPath

Author: @nicuparente Did we close on this with the Shell folks yet?

>         // Initialize using Shell assets like DisplayName and iconPath

Author: For now, I went ahead and removed the overload that takes in the color. I believe we are going with the theme color for the device as the POR unless we are told otherwise.

>         ToastAssets(String displayName, String iconPath);

Author: No more ToastAssets per feedback

        // The App friendly DisplayName for the toast in Action Centre 

>         String DisplayName { get; };

Author: The idea here was to make it Read-Only after initializing the construct. There wasn't any reason to make it Write.

>         // The full file path for the icon image

Author: It should be a Uri. Can make that change. Also absolute and relative file paths should work for unpackaged apps. Atleast that was the intention. Should I reconsider?

    runtimeclass ToastActivationInfo
    {
        // Initialize using a manifest defined COM Activator Id. Only applicable to Packaged Win32 applications
        static ToastActivationInfo CreateToastActivationInfoFromGuid(Guid taskClsid);

        // Initialize using Toast Assets. Only applicable to Unpackaged Win32 applications which need to specify their own assets like DisplayName and Icon.
        static ToastActivationInfo CreateToastActivationInfoFromToastAssets(ToastAssets assets);

        // The CLSID associated with the Client COM server that Windows App SDK will activate
        Guid TaskClsid{ get; };

        // The Shell assets associated with the Unpackaged app
        ToastAssets Assets{ get; };
    };

    [experimental]
    // Event args for the Toast Activation
    runtimeclass ToastActivatedEventArgs
    {
        // Arguments from the invoked button. Empty for Default Activation with no launch args specified in payload.
        String ActivationArgs{ get; };

        // The data from the input elements of a toast like a TextBox
        Windows.Foundation.Collections.IMapView<String, String> UserInput{ get; };

>     };

Author: The pipeline for all Toast Activation events originate from the Shell. So there isn't any redirection logic in this case unlike Push scenarios which target the App (unless it is cloud Push which targets the shell directly)

    [experimental]
    // Toast Progress Data
    runtimeclass ToastProgressData
    {
        // Initializes a new Instance of ToastProgressData
        ToastProgressData();

        // Gets or sets the sequence number of this notification data.
        // When multiple ToastProgressData objects are received, the system displays the data with the greatest non-zero number. 
        UInt32 SequenceNumber;

        // Gets/Sets the value for the title. Binds to {progressTitle} in progress xml tag.
        String Title;

        // Gets/Sets the Value for the numerical Progress percentile: a number between 0 and 1. Binds to {progressValue} in progress xml tag.
        Double Value;

        // Gets/Sets the Value for the Progress String. Binds to {progressValueString} in progress xml tag
        String ValueStringOverride;

        // Gets/Sets the Value for the Status. Binds to {progressStatus} in progress xml tag
        String Status;
    };

    [experimental]
    [flags]
    // The Toast User Setting or Toast Group Policy Setting
    enum ToastNotificationSetting
    {

>         Enabled, // Toast is not blocked by settings or group policy

Author: Agreed none of these are flags. We don't plan on any BIT operations here and each enum is mutually exclusive to the other.

>         Enabled, // Toast is not blocked by settings or group policy

Author: I'll remove the flag keyword

        DisabledForApplication, // Toast is blocked by a user defined App Setting
        DisabledForUser, // Toast is blocked by a user defined Global Setting
        DisabledByGroupPolicy, // Toast is blocked by Group Policy
        DisabledByManifest, // Toast is blocked by a setting in the manifest. Only for packaged applications.
    };

    [experimental]
    [flags]
    // Some basic predefined Toast Payload Templates

>     enum ToastTemplateType

Author: No they aren't. Removed.

    {
        ToastImageAndText01,
        ToastImageAndText02,
        ToastImageAndText03,
        ToastImageAndText04,
        ToastText01,
        ToastText02,
        ToastText03,
        ToastText04,
    };

    [experimental]
    [flags]
    // The Result for a Toast Progress related operation
    enum ToastProgressResult
    {
        Succeeded, // The progress operation succeeded
        Failed, // The progress operation failed
        NotificationNotFound, // The progress operation failed to find a toast to process updates
    };

    [experimental]

>     [flags]

Author: Removed

    enum ToastPriority
    {
        Default, // The notification should have default behavior in terms of delivery and display priority during connected standby mode.
        High, // The notification should be treated as high priority. For desktop PCs, this means during connected standby mode the incoming notification can turn on the screen for Surface-like devices if it doesn't have a closed lid detected.
    };

    [experimental]
    // Represent a toast Notification construct
    runtimeclass ToastNotification
    {
        // Initialize a new Toast using an XML Payload.
        ToastNotification(Windows.Data.Xml.Dom.XmlDocument payload);

        // Unique identifier used to replace a notification within a group.
        String Tag;

        // Unique identifier for a toast group in the app
        String Group;

        // A unique identifier for the Toast generated by the platform.

>         UInt32 ToastIdentifier;

Reviewer 13: Should this be {get} only? Why a developer could set its own toastId?

>         UInt32 ToastIdentifier;

Author: @danielayala94 Developer doesn't set the ToastId. But when they call ToastManager::Show, the SDK is supposed to populate the Id from the platform. Hence this is a get/set operation. Does that align with your expectations?

>         UInt32 ToastIdentifier;

Author: This one was resolved to be only a getter since the platform sets the id always! Not the developer.

        // The notification Xml Payload
        Windows.Data.Xml.Dom.XmlDocument Payload{ get; };

        // Gets or sets additional information about the toast progress.
        ToastProgressData ProgressData;

        // Gets or sets the time after which a toast notification should not be displayed.
        Windows.Foundation.DateTime ExpirationTime;

        // Indicates whether the toast will remain in the Action Center after a reboot.
        Boolean ExpiresOnReboot;

        // Gets or sets the priority for a Toast.
        // Hints on how and at what urgency level a notification should be presented to the user (whether to wake up the screen, etc).
        ToastPriority Priority;

        // Gets or sets whether a toast's pop-up UI is displayed on the user's screen.

>         Boolean SuppressPopup;

Reviewer 8: Or perhaps a bit more clearer `IsVisible`?

>         Boolean SuppressPopup;

Author: Going with SuppressDisplay for now since this is technically still visible in the actionCentre even after the value is set to true. It's just that the Popup won't show at the bottom right corner of the screen and its a silent action centre update.

>         Boolean SuppressPopup;

Reviewer 2: Oh, I didn't realize that SuppressDisplay=true still showed it in action center. I thought it suppressed it from displaying anywhere, like a raw notification.

Boolean ShowAsPopup; // default true

>         Boolean SuppressPopup;

Author: No. The popup won't show if it is set to false and the screen will still be turned off

    };

    [experimental]
    // Supports Toast related operations for all prior displayed Toasts in Action Centre
    runtimeclass ToastNotificationHistory
    {
        // Removes a specific toast with a specific toastIdentifier from Action Centre
        Windows.Foundation.IAsyncAction RemoveWithIdentiferAsync(UInt32 toastIdentifier);

        // Removes a toast having a specific tag
        Windows.Foundation.IAsyncAction RemoveWithTagAsync(String tag);

        // Removes a toast having a specific tag and group
        Windows.Foundation.IAsyncAction RemoveWithTagGroupAsync(String tag, String group);

        // Remove all toasts for a specific group
        Windows.Foundation.IAsyncAction RemoveGroupAllAsync(String group);

        // Removes all the toasts for the App from Action Centre
        Windows.Foundation.IAsyncAction RemoveAllAsync();

        // Gets all the toasts for the App from Action Centre
        Windows.Foundation.IAsyncOperation<Windows.Foundation.Collections.IVectorView<ToastNotification> > GetAllAsync();
    };

    [experimental]
    // The manager class which encompasses all Toast API Functionality

>     static runtimeclass ToastNotificationManager

Author: That sounds like a great idea. It should solve the static bloat problem in our *Manager classes. Perhaps do the same for Push as well.

    {
        // Register an activator using an ActivationInfo context and caches the token for unregister

>         static void RegisterActivator(ToastActivationInfo details);

Author: Correct. For Unregister to work, Register should have been called atleast once before after the app launched.

        // Unregisters the activator and removes the cached Registration token.
        static void UnregisterActivator();

        // Event handler for Toast Activations
        static event Windows.Foundation.EventHandler<ToastActivatedEventArgs> ToastActivated;

        // Displays the Toast in Action Centre
        static void ShowToast(ToastNotification toast);

        // Updates the Toast for a Progress related operation using Tag and Group
        static Windows.Foundation.IAsyncOperation<ToastProgressResult> UpdateToastProgressDataAsync(ToastProgressData data, String tag, String group);

        // Updates the Toast for a Progress related operation using Tag
        static Windows.Foundation.IAsyncOperation<ToastProgressResult> UpdateToastProgressDataAsync(ToastProgressData data, String tag);

        // Get the Toast Setting status for the app

>         static Windows.Foundation.IAsyncOperation<ToastNotificationSetting> Setting{ get; };

Author: Will remove the IAsyncOperation here

Toast messages could have a local or cloud based origin. In the case of local toasts, the message always originates from the app installed on the device. For cloud toasts, the message always originates from a remote application service that targets the locally installed app in question. 

Here is a visual representation of a Simple Toast with no interactive controls:

![Toast Simple](https://docs.microsoft.com/en-us/windows/apps/design/shell/tiles-and-notifications/images/send-toast-02.png)

Here is a visual representation of a Toast with simple button controls:

![Toast Interactive 1](https://docs.microsoft.com/en-us/windows/apps/design/shell/tiles-and-notifications/images/adaptivetoasts-structure.jpg)

Here is a visual representation of a Toast with a Message "Reply To" option:

![Toast Interactive 2](https://docs.microsoft.com/en-us/windows/apps/design/shell/tiles-and-notifications/images/toast-notification.png)

For more details see:

-   [Toast Notification WinRT APIs](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Notifications.ToastNotification?view=winrt-20348)

>     Defines all the API constructs that we have for Toast Notifications in WinRT today using the Legacy Windows SDK.

Author: Thats a typo. Its just Windows SDK. I'm going to fix that in the next iteration.

(MSIX Desktop Bridge, MSIX Win32App, Sparse Signed Packages). Moreover, all Toast scenarios should
adhere to OS resource management policies like Power Saver, Network Attribution (amount of data an
app uses), enterprise group policies, etc. The Windows App SDK Push component will abstract away the
complexities of dealing with Toast delivery and toast related activations as much as
possible freeing the developer to focus on other app related challenges.

We will prioritize the following feature set for Toast Notifications in Windows App SDK:

-   Toast Registrations for Unpackaged apps.
-   Toast Registrations for Packaged apps
-   Cloud sourced Toasts for Unpackaged Win32 apps.
-   Cloud sourced Toasts for Packaged Win32 apps.
-   Local Toast delivery and CRUD operations for Packaged Win32 apps.
-   Local Toasts delivery and CRUD operations for unpackaged Win32 apps.

The following feature set will not be prioritized for the next release but will be prioritized for the second half of 2022:

-   Object Model to compose Toasts without having to rely on Xml templates.

> -   Support for Granular toasts and toast collections.

Author: @nicuparente ? Can you help answer this? Should i remove it from the spec?

> -   Support for Granular toasts and toast collections.

Author: @nicuparente @danielayala94  I am removing references to Granular notifications. Will just call it Collections and end it at there.

    ToastNotificationManager::RegisterActivator(activationInfo);

    auto args = AppInstance::GetCurrent().GetActivatedEventArgs();
    auto kind = args.Kind();

    if (kind == ExtendedActivationKind::Launch)
    {
        // App is launched in FG. So intercept toast activators via FG event
        const auto token = ToastNotificationManager::ToastActivated([](const auto&, ToastActivatedEventArgs const& toastActivatedEventArgs)
        {
                ProcessToastArgs(toastActivatedEventArgs);
        });

        // App does Foreground Stuff Here

        // Cleanup
        ToastNotificationManager::ToastActivated(token);
    }
    else if (kind == ExtendedActivationKind::ToastNotification)
    {
        auto toastActivatedEventArgs = args.Data().as<ToastActivatedEventArgs>();
        ProcessToastArgs(toastActivatedEventArgs);
    }

    return 0;
}
```

## Unpackaged app Toast Registration

Similarly unpackaged apps can also Register themselves to send and act as targets for toast notifications.

```cpp
int main()
{
    ToastAssets assets = winrt::make<ToastAssets>(
        L"SampleApp",

>         L"c:\test\icon.png",

Author: Good catch

}
```

## Displaying a Toast

The app will need to display a toast to the user. To do this we recommend that the developer always sets a tag and group field so that similar toasts can replace it in the future. They also have the option to delete the toast automatically from ActionCentre on subsequent reboots by setting the ExpiresOnReboot property to true.

```cpp
// Display a toast with tag and group
void DisplayToast(const winrt::Windows::Data::Xml::Dom::XmlDocument& doc)
{
    ToastNotification toast = winrt::make<ToastNotification>(doc);
    // Setting Tag, Group and ExpiresOnReboot properties are optional
    toast.Tag(L"Tag");
    toast.Group(L"Group");
    toast.ExpiresOnReboot(true);
    ToastNotificationManager::ShowToast(toast);
}
```

## Create a Toast Payload using Preexisting Toast XML templates

We will provide a set of templates that developers can use to compose payloads. The developer can then use the Document Object Model (DOM) to customize the content in the payload. The templates that we are introducing are as follows:

| Template Name | Scenario Examples | Example |
| :--- | :---| :--- |
|BasicToast| Simple alerts | ![BasicToast](images/BasicToast.png) |

> |ToastWithTwoButtons| Quick actions (2 options) such as view more information or remind me later | ![BasicToast](images/ToastWithTwoButtons.png) |

Author: Should we have a toast with one button? Or do we think thats not critical?

Author: @vaheeshta

Reviewer 12: Shell believes not critical, but toast with icons is, so added that.

Author: Thanks @vaheeshta .

                                            L"status = \"{progressStatus}\" />"
                                    L"</binding>"
                                L"</visual>"
                            L"</toast>";

    winrt::Windows::Data::Xml::Dom::XmlDocument doc = winrt::make<winrt::Windows::Data::Xml::Dom::XmlDocument>();
    doc.LoadXml(payload);
    ToastNotification toast = winrt::make<ToastNotification>(doc);
    toast.Tag(tag);
    toast.Group(group);

    // Assign initial values for first toast progress UI
    ToastProgressData data = winrt::make<ToastProgressData>();
    data.Title(L"Weekly playlist"); // Binds to {progressTitle} in xml payload
    data.Value(0.6); // Binds to {progressValue} in xml payload
    data.ValueStringOverride(L"15/26 songs"); // Binds to {progressValueString} in xml payload
    data.Status(L"Downloading..."); // Binds to {progressStatus} in xml payload
    data.SequenceNumber(1); // The sequence number has to be incremental for the platform to render progress

    toast.ProgressData(data);
    ToastNotificationManager::ShowToast(toast);
}

// Send subsequent progress updates
winrt::Windows::Foundation::IAsyncAction UpdateProgressAsync()
{

>     winrt::hstring tag = L"weekly-playlist";

Reviewer 13: Maybe we should advice the devs to have such strings as member values? Otherwise, there is a chance to get it wrong.

Author: I grow weary of that. From my point of view, its purely an implementation detail. I would argue that they should be arguments to this class but i specified it as a local declaration to get the point across on purpose.

        ToastWithAvatarAndDropdown,
        ToastWithDropdown,
        ToastWithReminder,
        ToastWithAlarm,
        ToastWithHeroImage,
        ToastWithInlineImage,
    };

    [experimental]
    [flags]
    // The Result for a Toast Progress related operation
    enum ToastProgressResult
    {
        Succeeded, // The progress operation succeeded
        Failed, // The progress operation failed
        NotificationNotFound, // The progress operation failed to find a toast to process updates
    };

    [experimental]
    [flags]
    enum ToastPriority
    {
        Default, // The notification should have default behavior in terms of delivery and display priority during connected standby mode.
        High, // The notification should be treated as high priority. For desktop PCs, this means during connected standby mode the incoming notification can turn on the screen for Surface-like devices if it doesn't have a closed lid detected.
    };

    [experimental]
    // Represent a toast Notification construct
    runtimeclass ToastNotification
    {
        // Initialize a new Toast using an XML Payload.
        ToastNotification(Windows.Data.Xml.Dom.XmlDocument payload);

>         // Unique identifier used to replace a notification within a group.

Reviewer 13: Technically the group isn't needed. With a tag you can still do things like remove toast.

Author: True. you can get away with replacement of just a tag. But you can also have a tag within a group.

    }

    // other app init and then message loop here

    // Cleanup event handler
    manager.NotificationInvoked(token);

    // Call Unregister() before exiting main so that subsequent invocations will launch a new process
    manager.Unregister();
    return 0;
}
```

## Displaying an App Notification

> To display a Notification, an app needs to define a payload using in xml. In the example below, the

Reviewer 3: "using in xml" -> "in xml".

Author: oops. Thanks for catching this.
